

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>openff.toolkit.utils.openeye_wrapper &mdash; swiftpol  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../../_static/css/theme.css?v=e59714d7" />

  
      <script src="../../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../../index.html" class="icon icon-home">
            swiftpol
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../getting_started.html#getting-started-with-swiftpol">Getting Started with Swiftpol</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#module-swiftpol.build">swiftpol.build</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../api.html#swiftpol-parameterize">swiftpol.parameterize</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">swiftpol</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">openff.toolkit.utils.openeye_wrapper</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for openff.toolkit.utils.openeye_wrapper</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Wrapper class providing a minimal consistent interface to</span>
<span class="sd">the `OpenEye Toolkit &lt;https://docs.eyesopen.com/toolkits/python/quickstart-python/index.html&gt;`_</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">__all__</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;OpenEyeToolkitWrapper&quot;</span><span class="p">,)</span>


<span class="c1"># See https://github.com/conda-forge/openff-toolkit-feedstock/issues/86</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">zstandard</span>  <span class="c1"># noqa</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="k">pass</span>
<span class="kn">import</span> <span class="nn">importlib</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">wraps</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">TYPE_CHECKING</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">cachetools</span> <span class="kn">import</span> <span class="n">LRUCache</span><span class="p">,</span> <span class="n">cached</span>
<span class="kn">from</span> <span class="nn">typing_extensions</span> <span class="kn">import</span> <span class="n">TypeAlias</span>

<span class="kn">from</span> <span class="nn">openff.toolkit</span> <span class="kn">import</span> <span class="n">Quantity</span><span class="p">,</span> <span class="n">unit</span>

<span class="k">if</span> <span class="n">TYPE_CHECKING</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">openff.toolkit.topology.molecule</span> <span class="kn">import</span> <span class="n">Atom</span><span class="p">,</span> <span class="n">Bond</span><span class="p">,</span> <span class="n">FrozenMolecule</span><span class="p">,</span> <span class="n">Molecule</span>

<span class="kn">from</span> <span class="nn">openff.units.elements</span> <span class="kn">import</span> <span class="n">SYMBOLS</span>

<span class="kn">from</span> <span class="nn">openff.toolkit.utils.base_wrapper</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">ToolkitWrapper</span><span class="p">,</span>
    <span class="n">_ChargeSettings</span><span class="p">,</span>
    <span class="n">_mol_to_ctab_and_aro_key</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">openff.toolkit.utils.constants</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">ALLOWED_AROMATICITY_MODELS</span><span class="p">,</span>
    <span class="n">DEFAULT_AROMATICITY_MODEL</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">openff.toolkit.utils.exceptions</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">ChargeCalculationError</span><span class="p">,</span>
    <span class="n">ChargeMethodUnavailableError</span><span class="p">,</span>
    <span class="n">ConformerGenerationError</span><span class="p">,</span>
    <span class="n">EmptyInChiError</span><span class="p">,</span>
    <span class="n">GAFFAtomTypeWarning</span><span class="p">,</span>
    <span class="n">InChIParseError</span><span class="p">,</span>
    <span class="n">InconsistentStereochemistryError</span><span class="p">,</span>
    <span class="n">InvalidAromaticityModelError</span><span class="p">,</span>
    <span class="n">InvalidIUPACNameError</span><span class="p">,</span>
    <span class="n">LicenseError</span><span class="p">,</span>
    <span class="n">MultipleComponentsInMoleculeWarning</span><span class="p">,</span>
    <span class="n">NotAttachedToMoleculeError</span><span class="p">,</span>
    <span class="n">OpenEyeError</span><span class="p">,</span>
    <span class="n">OpenEyeImportError</span><span class="p">,</span>
    <span class="n">RadicalsNotSupportedError</span><span class="p">,</span>
    <span class="n">SMILESParseError</span><span class="p">,</span>
    <span class="n">ToolkitUnavailableException</span><span class="p">,</span>
    <span class="n">UnassignedChemistryInPDBError</span><span class="p">,</span>
    <span class="n">UndefinedStereochemistryError</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">openff.toolkit.utils.utils</span> <span class="kn">import</span> <span class="n">inherit_docstrings</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>


<span class="n">TTA</span><span class="p">:</span> <span class="n">TypeAlias</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="n">Any</span><span class="p">,</span> <span class="o">...</span><span class="p">],</span> <span class="o">...</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">get_oeformat</span><span class="p">(</span><span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

    <span class="n">file_format</span> <span class="o">=</span> <span class="n">file_format</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="c1"># XXX This is what RDKit does. Should be supported here too?</span>
    <span class="k">if</span> <span class="n">file_format</span> <span class="o">==</span> <span class="s2">&quot;MOL&quot;</span><span class="p">:</span>
        <span class="n">file_format</span> <span class="o">=</span> <span class="s2">&quot;SDF&quot;</span>

    <span class="n">oeformat</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">oechem</span><span class="p">,</span> <span class="s2">&quot;OEFormat_&quot;</span> <span class="o">+</span> <span class="n">file_format</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">oeformat</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Unsupported file format: </span><span class="si">{</span><span class="n">file_format</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">oeformat</span>


<span class="nd">@inherit_docstrings</span>
<span class="k">class</span> <span class="nc">OpenEyeToolkitWrapper</span><span class="p">(</span><span class="n">ToolkitWrapper</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    OpenEye toolkit wrapper</span>

<span class="sd">    .. warning :: This API is experimental and subject to change.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_toolkit_name</span> <span class="o">=</span> <span class="s2">&quot;OpenEye Toolkit&quot;</span>
    <span class="n">_toolkit_installation_instructions</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;The OpenEye Toolkits can be installed via &quot;</span>
        <span class="s2">&quot;`mamba install openeye-toolkits -c openeye`&quot;</span>
    <span class="p">)</span>
    <span class="n">_toolkit_license_instructions</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;The OpenEye Toolkits require a (free for academics) license, see &quot;</span>
        <span class="s2">&quot;https://docs.eyesopen.com/toolkits/python/quickstart-python/license.html&quot;</span>
    <span class="p">)</span>
    <span class="c1"># This could belong to ToolkitWrapper, although it seems strange</span>
    <span class="c1"># to carry that data for open-source toolkits</span>
    <span class="n">_is_licensed</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="c1"># Only for OpenEye is there potentially a difference between</span>
    <span class="c1"># being available and installed</span>
    <span class="n">_is_installed</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">_license_functions</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;oechem&quot;</span><span class="p">:</span> <span class="s2">&quot;OEChemIsLicensed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;oequacpac&quot;</span><span class="p">:</span> <span class="s2">&quot;OEQuacPacIsLicensed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;oeiupac&quot;</span><span class="p">:</span> <span class="s2">&quot;OEIUPACIsLicensed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;oeomega&quot;</span><span class="p">:</span> <span class="s2">&quot;OEOmegaIsLicensed&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">_supported_charge_methods</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;am1bcc&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEAM1BCCCharges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;am1-mulliken&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEAM1Charges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;gasteiger&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEGasteigerCharges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;mmff94&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEMMFF94Charges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;am1bccnosymspt&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEAM1BCCCharges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;am1elf10&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEELFCharges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># type: ignore[typeddict-item]</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">500</span><span class="p">,</span>
        <span class="p">},</span>
        <span class="s2">&quot;am1bccelf10&quot;</span><span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;oe_charge_method&quot;</span><span class="p">:</span> <span class="s2">&quot;OEAM1BCCELF10Charges&quot;</span><span class="p">,</span>
            <span class="s2">&quot;min_confs&quot;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
            <span class="s2">&quot;max_confs&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>  <span class="c1"># type: ignore[typeddict-item]</span>
            <span class="s2">&quot;rec_confs&quot;</span><span class="p">:</span> <span class="mi">500</span><span class="p">,</span>
        <span class="p">},</span>
    <span class="p">}</span>

    <span class="n">SUPPORTED_CHARGE_METHODS</span> <span class="o">=</span> <span class="n">_supported_charge_methods</span>

    <span class="n">_toolkit_file_read_formats</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;CAN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CDX&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CSV&quot;</span><span class="p">,</span>
        <span class="s2">&quot;FASTA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;INCHI&quot;</span><span class="p">,</span>
        <span class="s2">&quot;INCHIKEY&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ISM&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MDL&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MMOD&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MOL2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MOL2H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MOPAC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;OEB&quot;</span><span class="p">,</span>
        <span class="s2">&quot;PDB&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RDF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SDF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SKC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SLN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SMI&quot;</span><span class="p">,</span>
        <span class="s2">&quot;USM&quot;</span><span class="p">,</span>
        <span class="s2">&quot;XYC&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">_toolkit_file_write_formats</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;CAN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CDX&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CSV&quot;</span><span class="p">,</span>
        <span class="s2">&quot;FASTA&quot;</span><span class="p">,</span>
        <span class="s2">&quot;INCHI&quot;</span><span class="p">,</span>
        <span class="s2">&quot;INCHIKEY&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ISM&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MDL&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MMOD&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MOL2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MOL2H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;MOPAC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;OEB&quot;</span><span class="p">,</span>
        <span class="s2">&quot;PDB&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RDF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SDF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SKC&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SLN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SMI&quot;</span><span class="p">,</span>
        <span class="s2">&quot;USM&quot;</span><span class="p">,</span>
        <span class="s2">&quot;XYC&quot;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="c1"># check if the toolkit can be loaded</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_available</span><span class="p">():</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_is_installed</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">ToolkitUnavailableException</span><span class="p">(</span>
                    <span class="s2">&quot;OpenEye Toolkits are not installed.&quot;</span>
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_toolkit_installation_instructions</span>
                <span class="p">)</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_is_licensed</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">LicenseError</span><span class="p">(</span>
                    <span class="s2">&quot;The OpenEye Toolkits are found to be installed but not licensed and &quot;</span>
                    <span class="o">+</span> <span class="s2">&quot;therefore will not be used.</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_toolkit_license_instructions</span>
                <span class="p">)</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">__version__</span> <span class="k">as</span> <span class="n">openeye_version</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_toolkit_version</span> <span class="o">=</span> <span class="n">openeye_version</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_check_licenses</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Check license of all known OpenEye tools. Returns True if any are found</span>
<span class="sd">        to be licensed, False if all are not.&quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">tool</span><span class="p">,</span> <span class="n">license_func</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_license_functions</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">module</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="s2">&quot;openeye.&quot;</span> <span class="o">+</span> <span class="n">tool</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">module</span><span class="p">,</span> <span class="n">license_func</span><span class="p">)():</span>
                    <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">is_available</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Check if the given OpenEye toolkit components are available.</span>

<span class="sd">        If the OpenEye toolkit is not installed or no license is found</span>
<span class="sd">        for at least one the required toolkits , ``False`` is returned.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        all_installed</span>
<span class="sd">            ``True`` if all required OpenEye tools are installed and licensed,</span>
<span class="sd">            ``False`` otherwise</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_available</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_licensed</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">cls</span><span class="o">.</span><span class="n">_is_licensed</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_check_licenses</span><span class="p">()</span>
            <span class="k">if</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_installed</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">tool</span> <span class="ow">in</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_license_functions</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">_is_installed</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="s2">&quot;openeye.&quot;</span> <span class="o">+</span> <span class="n">tool</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
                        <span class="bp">cls</span><span class="o">.</span><span class="n">_is_installed</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_installed</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_licensed</span><span class="p">:</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">_is_available</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">cls</span><span class="o">.</span><span class="n">_is_available</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="bp">cls</span><span class="o">.</span><span class="n">_is_available</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_installed</span> <span class="ow">and</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_licensed</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_is_available</span>  <span class="c1"># type: ignore[return-value]</span>

    <span class="k">def</span> <span class="nf">from_object</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">obj</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Convert an OEMol (or OEMol-derived object) into an openff.toolkit.topology.molecule</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        obj</span>
<span class="sd">            An object to by type-checked.</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            Whether to accept molecules with undefined stereocenters. If False,</span>
<span class="sd">            an exception will be raised if a molecule with undefined stereochemistry</span>
<span class="sd">            is passed into this function.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        Molecule</span>
<span class="sd">            An openff.toolkit.topology.molecule Molecule.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        NotImplementedError</span>
<span class="sd">            If the object could not be converted into a Molecule.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># TODO: Add tests for the from_object functions</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">if</span> <span class="n">_cls</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="kn">from</span> <span class="nn">openff.toolkit.topology.molecule</span> <span class="kn">import</span> <span class="n">Molecule</span>

            <span class="n">_cls</span> <span class="o">=</span> <span class="n">Molecule</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">obj</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMolBase</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
                <span class="n">oemol</span><span class="o">=</span><span class="n">obj</span><span class="p">,</span>
                <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="n">allow_undefined_stereo</span><span class="p">,</span>
                <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Cannot create Molecule from </span><span class="si">{</span><span class="nb">type</span><span class="p">(</span><span class="n">obj</span><span class="p">)</span><span class="si">}</span><span class="s2"> object&quot;</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_polymer_openmm_topology_to_offmol</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">mol_class</span><span class="p">,</span> <span class="n">omm_top</span><span class="p">,</span> <span class="n">substructure_dictionary</span>
    <span class="p">):</span>
        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_polymer_openmm_topology_to_oemol</span><span class="p">(</span><span class="n">omm_top</span><span class="p">,</span> <span class="n">substructure_dictionary</span><span class="p">)</span>
        <span class="n">offmol</span> <span class="o">=</span> <span class="n">mol_class</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">offmol</span>

    <span class="k">def</span> <span class="nf">_polymer_openmm_topology_to_oemol</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">omm_top</span><span class="p">,</span>
        <span class="n">substructure_library</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        omm_top</span>
<span class="sd">            OpenMM Topology loaded from PDB</span>
<span class="sd">        substructure_library</span>
<span class="sd">            A dictionary of substructures. substructure_library[aa_name] = list[tagged SMARTS, list[atom_names]]</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        oemol</span>
<span class="sd">            a new molecule with charges and bond order added</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_connectivity_from_openmm_top</span><span class="p">(</span><span class="n">omm_top</span><span class="p">)</span>

        <span class="n">already_assigned_nodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">already_assigned_edges</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="c1"># Keeping track of which atoms are matched where will help us with error</span>
        <span class="c1"># messages</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">res_name</span> <span class="ow">in</span> <span class="n">substructure_library</span><span class="p">:</span>
            <span class="c1"># TODO: This is a hack for the moment since we don&#39;t have a more sophisticated way to resolve clashes</span>
            <span class="c1"># so it just does the biggest substructures first</span>
            <span class="n">sorted_substructure_smarts</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                <span class="n">substructure_library</span><span class="p">[</span><span class="n">res_name</span><span class="p">],</span> <span class="n">key</span><span class="o">=</span><span class="nb">len</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">substructure_smarts</span> <span class="ow">in</span> <span class="n">sorted_substructure_smarts</span><span class="p">:</span>
                <span class="n">ss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fuzzy_query</span><span class="p">(</span><span class="n">substructure_smarts</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">ss</span><span class="o">.</span><span class="n">Match</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="kc">True</span><span class="p">):</span>
                    <span class="n">match_mol_atom_indices</span> <span class="o">=</span> <span class="p">[</span>
                        <span class="n">at</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="k">for</span> <span class="n">at</span> <span class="ow">in</span> <span class="n">match</span><span class="o">.</span><span class="n">GetTargetAtoms</span><span class="p">()</span>
                    <span class="p">]</span>
                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">match_mol_atom_indices</span><span class="p">:</span>
                        <span class="n">matches</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">res_name</span><span class="p">)</span>

                    <span class="k">if</span> <span class="nb">any</span><span class="p">(</span>
                        <span class="n">m</span> <span class="ow">in</span> <span class="n">already_assigned_nodes</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">match_mol_atom_indices</span>
                    <span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">res_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;PEPTIDE_BOND&quot;</span><span class="p">,</span> <span class="s2">&quot;DISULFIDE&quot;</span><span class="p">]):</span>
                        <span class="k">continue</span>

                    <span class="k">for</span> <span class="n">substructure_atom</span><span class="p">,</span> <span class="n">mol_atom</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                        <span class="k">match</span><span class="o">.</span><span class="n">GetPatternAtoms</span><span class="p">(),</span> <span class="n">match</span><span class="o">.</span><span class="n">GetTargetAtoms</span><span class="p">()</span>
                    <span class="p">):</span>
                        <span class="n">mol_atom</span><span class="o">.</span><span class="n">SetFormalCharge</span><span class="p">(</span><span class="n">substructure_atom</span><span class="o">.</span><span class="n">GetFormalCharge</span><span class="p">())</span>
                        <span class="c1"># Set arbitrary initial stereochemistry to avoid</span>
                        <span class="c1"># spamming &quot;undefined stereo&quot; warnings. In the from_polymer_pdb</span>
                        <span class="c1"># code path, the &quot;real stereo&quot; will be assigned later by a</span>
                        <span class="c1"># call to _assign_aromaticity_and_stereo_from_3d.</span>
                        <span class="n">neighs</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">mol_atom</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()]</span>
                        <span class="n">mol_atom</span><span class="o">.</span><span class="n">SetStereo</span><span class="p">(</span>
                            <span class="n">neighs</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomStereo_Tetra</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomStereo_Left</span>
                        <span class="p">)</span>

                        <span class="n">already_assigned_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mol_atom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">())</span>
                    <span class="k">for</span> <span class="n">substructure_bond</span><span class="p">,</span> <span class="n">mol_bond</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span>
                        <span class="k">match</span><span class="o">.</span><span class="n">GetPatternBonds</span><span class="p">(),</span> <span class="n">match</span><span class="o">.</span><span class="n">GetTargetBonds</span><span class="p">()</span>
                    <span class="p">):</span>
                        <span class="n">mol_bond</span><span class="o">.</span><span class="n">SetOrder</span><span class="p">(</span><span class="n">substructure_bond</span><span class="o">.</span><span class="n">GetOrder</span><span class="p">())</span>
                        <span class="n">already_assigned_edges</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
                            <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">mol_bond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">(),</span> <span class="n">mol_bond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">()]))</span>
                        <span class="p">)</span>

        <span class="n">oemol_n_atoms</span> <span class="o">=</span> <span class="nb">len</span><span class="p">([</span><span class="o">*</span><span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()])</span>
        <span class="n">unassigned_atoms</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">oemol_n_atoms</span><span class="p">))</span> <span class="o">-</span> <span class="n">already_assigned_nodes</span><span class="p">)</span>

        <span class="n">all_bonds</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">bond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">(),</span> <span class="n">bond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">()]))</span>
                <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">()</span>
            <span class="p">]</span>
        <span class="p">)</span>
        <span class="n">unassigned_bonds</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">all_bonds</span> <span class="o">-</span> <span class="n">already_assigned_edges</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">unassigned_atoms</span> <span class="ow">or</span> <span class="n">unassigned_bonds</span><span class="p">:</span>
            <span class="c1"># Some advanced error reporting needs to interpret the substructure smarts to do things like</span>
            <span class="c1"># compare atom counts. Since OFFTK doesn&#39;t have a native class to hold fragments, we convert</span>
            <span class="c1"># the smarts into a sorted list of symbols to help with generating the error message.</span>
            <span class="n">resname_to_symbols_and_atomnames</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">resname</span><span class="p">,</span> <span class="n">smarts_to_atom_names</span> <span class="ow">in</span> <span class="n">substructure_library</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">resname_to_symbols_and_atomnames</span><span class="p">[</span><span class="n">resname</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">smarts</span><span class="p">,</span> <span class="n">atom_names</span> <span class="ow">in</span> <span class="n">smarts_to_atom_names</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">qmol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEQMol</span><span class="p">()</span>
                    <span class="n">oechem</span><span class="o">.</span><span class="n">OEParseSmiles</span><span class="p">(</span><span class="n">qmol</span><span class="p">,</span> <span class="n">smarts</span><span class="p">)</span>
                    <span class="n">symbols</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
                        <span class="p">[</span><span class="n">SYMBOLS</span><span class="p">[</span><span class="n">atom</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()]</span> <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">qmol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()]</span>
                    <span class="p">)</span>
                    <span class="n">resname_to_symbols_and_atomnames</span><span class="p">[</span><span class="n">resname</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="p">(</span><span class="n">symbols</span><span class="p">,</span> <span class="n">atom_names</span><span class="p">)</span>
                    <span class="p">)</span>

            <span class="k">raise</span> <span class="n">UnassignedChemistryInPDBError</span><span class="p">(</span>
                <span class="n">substructure_library</span><span class="o">=</span><span class="n">resname_to_symbols_and_atomnames</span><span class="p">,</span>
                <span class="n">omm_top</span><span class="o">=</span><span class="n">omm_top</span><span class="p">,</span>
                <span class="n">unassigned_atoms</span><span class="o">=</span><span class="n">unassigned_atoms</span><span class="p">,</span>
                <span class="n">unassigned_bonds</span><span class="o">=</span><span class="n">unassigned_bonds</span><span class="p">,</span>
                <span class="n">matches</span><span class="o">=</span><span class="n">matches</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">oemol</span>

    <span class="k">def</span> <span class="nf">_get_connectivity_from_openmm_top</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">omm_top</span><span class="p">):</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">()</span>
        <span class="c1"># Add atoms</span>
        <span class="n">oemol_atoms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>  <span class="c1"># list of corresponding oemol atoms</span>
        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">omm_top</span><span class="o">.</span><span class="n">atoms</span><span class="p">():</span>
            <span class="n">oeatom</span> <span class="o">=</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NewAtom</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">element</span><span class="o">.</span><span class="n">atomic_number</span><span class="p">)</span>
            <span class="n">oemol_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span>

        <span class="c1"># Add bonds</span>
        <span class="n">oemol_bonds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>  <span class="c1"># list of corresponding oemol bonds</span>
        <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">omm_top</span><span class="o">.</span><span class="n">bonds</span><span class="p">():</span>
            <span class="n">atom1_index</span> <span class="o">=</span> <span class="n">bond</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
            <span class="n">atom2_index</span> <span class="o">=</span> <span class="n">bond</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
            <span class="n">oebond</span> <span class="o">=</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NewBond</span><span class="p">(</span><span class="n">oemol_atoms</span><span class="p">[</span><span class="n">atom1_index</span><span class="p">],</span> <span class="n">oemol_atoms</span><span class="p">[</span><span class="n">atom2_index</span><span class="p">])</span>
            <span class="n">oemol_bonds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oebond</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">oemol</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_fuzzy_query</span><span class="p">(</span><span class="n">query</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;return a copy of Query which is less specific:</span>
<span class="sd">        - ignore aromaticity and hybridization of atoms (i.e. [#6] not C)</span>
<span class="sd">        - ignore bond orders</span>
<span class="sd">        - ignore formal charges</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="kn">from</span> <span class="nn">openff.toolkit.utils.exceptions</span> <span class="kn">import</span> <span class="n">SMIRKSParsingError</span>

        <span class="c1">#  Jeff wasn&#39;t able to get this working with OEQMol and OEParseSmarts,</span>
        <span class="c1">#  the QMol/SS matching didn&#39;t behave correctly when set to AtomicNumber</span>
        <span class="n">qmol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">()</span>
        <span class="n">status</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEParseSmiles</span><span class="p">(</span>
            <span class="n">qmol</span><span class="p">,</span>
            <span class="n">query</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">status</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">SMIRKSParsingError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;OpenEye Toolkit was unable to parse SMIRKS </span><span class="si">{</span><span class="n">query</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="n">ss</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESubSearch</span><span class="p">(</span><span class="n">qmol</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEExprOpts_AtomicNumber</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ss</span>

    <span class="k">def</span> <span class="nf">from_file</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">file_path</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">],</span>
        <span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return an openff.toolkit.topology.Molecule from a file using this toolkit.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        file_path</span>
<span class="sd">            The file to read the molecule from</span>
<span class="sd">        file_format</span>
<span class="sd">            Format specifier, usually file suffix (eg. &#39;MOL2&#39;, &#39;SMI&#39;)</span>
<span class="sd">            Note that not all toolkits support all formats. Check ToolkitWrapper.toolkit_file_read_formats for details.</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            If false, raises an exception if oemol contains undefined stereochemistry.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecules</span>
<span class="sd">            The list of ``Molecule`` objects in the file.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        GAFFAtomTypeWarning</span>
<span class="sd">            If the loaded mol2 file possibly uses GAFF atom types, which</span>
<span class="sd">            are not supported.</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        Load a mol2 file into an OpenFF ``Molecule`` object.</span>

<span class="sd">        &gt;&gt;&gt; from openff.toolkit.utils import get_data_file_path</span>
<span class="sd">        &gt;&gt;&gt; mol2_file_path = get_data_file_path(&#39;molecules/cyclohexane.mol2&#39;)</span>
<span class="sd">        &gt;&gt;&gt; toolkit = OpenEyeToolkitWrapper()</span>
<span class="sd">        &gt;&gt;&gt; molecule = toolkit.from_file(mol2_file_path, file_format=&#39;mol2&#39;)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">):</span>
            <span class="n">_file_path</span> <span class="o">=</span> <span class="n">file_path</span><span class="o">.</span><span class="n">as_posix</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">_file_path</span> <span class="o">=</span> <span class="n">file_path</span>

        <span class="n">oeformat</span> <span class="o">=</span> <span class="n">get_oeformat</span><span class="p">(</span><span class="n">file_format</span><span class="p">)</span>
        <span class="n">ifs</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">oemolistream</span><span class="p">(</span><span class="n">_file_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ifs</span><span class="o">.</span><span class="n">IsValid</span><span class="p">():</span>
            <span class="c1"># Get Python to report an error message, if possible.</span>
            <span class="c1"># This can distinguish between FileNotFound, IsADirectoryError, etc.</span>
            <span class="nb">open</span><span class="p">(</span><span class="n">_file_path</span><span class="p">)</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="c1"># If that worked, then who knows. Fail anyway.</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="s2">&quot;Unable to open file&quot;</span><span class="p">)</span>

        <span class="n">ifs</span><span class="o">.</span><span class="n">SetFormat</span><span class="p">(</span><span class="n">oeformat</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_oemolistream_molecules</span><span class="p">(</span>
            <span class="n">ifs</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="p">,</span> <span class="n">file_path</span><span class="o">=</span><span class="n">_file_path</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">from_file_obj</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">file_obj</span><span class="p">,</span>
        <span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="s2">&quot;Molecule&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return an openff.toolkit.topology.Molecule from a file-like object (an object with a &quot;.read()&quot; method using</span>
<span class="sd">        this toolkit.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        file_obj</span>
<span class="sd">            The file-like object to read the molecule from</span>
<span class="sd">        file_format</span>
<span class="sd">            Format specifier, usually file suffix (eg. &#39;MOL2&#39;, &#39;SMI&#39;)</span>
<span class="sd">            Note that not all toolkits support all formats. Check ToolkitWrapper.toolkit_file_read_formats for details.</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            If false, raises an exception if oemol contains undefined stereochemistry.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecules</span>
<span class="sd">            The list of Molecule objects in the file object.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        GAFFAtomTypeWarning</span>
<span class="sd">            If the loaded mol2 file possibly uses GAFF atom types, which</span>
<span class="sd">            are not supported.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="c1"># Configure input molecule stream.</span>
        <span class="n">ifs</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">oemolistream</span><span class="p">()</span>
        <span class="n">ifs</span><span class="o">.</span><span class="n">openstring</span><span class="p">(</span><span class="n">file_obj</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
        <span class="n">oeformat</span> <span class="o">=</span> <span class="n">get_oeformat</span><span class="p">(</span><span class="n">file_format</span><span class="p">)</span>
        <span class="n">ifs</span><span class="o">.</span><span class="n">SetFormat</span><span class="p">(</span><span class="n">oeformat</span><span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_read_oemolistream_molecules</span><span class="p">(</span><span class="n">ifs</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">to_file_obj</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span> <span class="n">file_obj</span><span class="p">,</span> <span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes an OpenFF Molecule to a file-like object</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to write</span>
<span class="sd">        file_obj</span>
<span class="sd">            The file-like object to write to</span>
<span class="sd">        file_format</span>
<span class="sd">            The format for writing the molecule data</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># This function requires a text-mode file_obj.</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">file_obj</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
            <span class="c1"># Switch to a ValueError and use a more informative exception</span>
            <span class="c1"># message to match RDKit.</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Need a text mode file object like StringIO or a file opened with mode &#39;t&#39;&quot;</span>
            <span class="p">)</span> <span class="kn">from</span> <span class="kc">None</span>

        <span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryDirectory</span><span class="p">()</span> <span class="k">as</span> <span class="n">tmpdir</span><span class="p">:</span>
            <span class="n">path</span> <span class="o">=</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">(</span><span class="n">tmpdir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;input.</span><span class="si">{</span><span class="n">file_format</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">to_file</span><span class="p">(</span><span class="n">molecule</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">path</span><span class="p">),</span> <span class="n">file_format</span><span class="p">)</span>
            <span class="n">file_data</span> <span class="o">=</span> <span class="n">path</span><span class="o">.</span><span class="n">read_text</span><span class="p">()</span>
            <span class="n">file_obj</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">file_data</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">to_file</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">file_path</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">],</span>
        <span class="n">file_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes an OpenFF Molecule to a file-like object</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to write</span>
<span class="sd">        file_path</span>
<span class="sd">            The file path to write to.</span>
<span class="sd">        file_format</span>
<span class="sd">            The format for writing the molecule data</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">):</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">file_path</span><span class="o">.</span><span class="n">as_posix</span><span class="p">()</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
        <span class="n">ofs</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">oemolostream</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">ofs</span><span class="o">.</span><span class="n">IsValid</span><span class="p">():</span>
            <span class="c1"># Get Python to report an error message, if possible.</span>
            <span class="c1"># This can distinguish between PermissionError, IsADirectoryError, etc.</span>
            <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="c1"># If that worked, then who knows. Fail anyway.</span>
            <span class="k">raise</span> <span class="ne">OSError</span><span class="p">(</span><span class="s2">&quot;Unable to open file&quot;</span><span class="p">)</span>

        <span class="n">openeye_format</span> <span class="o">=</span> <span class="n">get_oeformat</span><span class="p">(</span><span class="n">file_format</span><span class="p">)</span>
        <span class="n">ofs</span><span class="o">.</span><span class="n">SetFormat</span><span class="p">(</span><span class="n">openeye_format</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">openeye_format</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEFormat_SMI</span><span class="p">:</span>
            <span class="n">ofs</span><span class="o">.</span><span class="n">SetFlavor</span><span class="p">(</span>
                <span class="n">openeye_format</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_get_smiles_flavor</span><span class="p">(</span><span class="n">isomeric</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">explicit_hydrogens</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
            <span class="p">)</span>

        <span class="c1"># OFFTK strictly treats SDF as a single-conformer format.</span>
        <span class="c1"># We need to override OETK&#39;s behavior here if the user is saving a multiconformer molecule.</span>

        <span class="c1"># Remove all but the first conformer when writing to SDF as we only support single conformer format</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">file_format</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;sdf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NumConfs</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">conf1</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">oemol</span><span class="o">.</span><span class="n">GetConfs</span><span class="p">()))</span>
            <span class="n">flat_coords</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">coord</span> <span class="ow">in</span> <span class="n">conf1</span><span class="o">.</span><span class="n">GetCoords</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">flat_coords</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">coord</span><span class="p">)</span>
            <span class="n">oemol</span><span class="o">.</span><span class="n">DeleteConfs</span><span class="p">()</span>
            <span class="n">oecoords</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEFloatArray</span><span class="p">(</span><span class="n">flat_coords</span><span class="p">)</span>
            <span class="n">oemol</span><span class="o">.</span><span class="n">NewConf</span><span class="p">(</span><span class="n">oecoords</span><span class="p">)</span>
        <span class="c1"># We&#39;re standardizing on putting partial charges into SDFs under the `atom.dprop.PartialCharge` property</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">file_format</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;sdf&quot;</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">):</span>
            <span class="n">partial_charges_list</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">oeatom</span><span class="o">.</span><span class="n">GetPartialCharge</span><span class="p">()</span> <span class="k">for</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()</span>
            <span class="p">]</span>
            <span class="n">partial_charges_str</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">val</span><span class="si">:</span><span class="s2">f</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">partial_charges_list</span><span class="p">])</span>
            <span class="c1"># TODO: &quot;dprop&quot; means &quot;double precision&quot; -- Is there any way to make Python more accurately</span>
            <span class="c1">#  describe/infer the proper data type?</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OESetSDData</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="s2">&quot;atom.dprop.PartialCharge&quot;</span><span class="p">,</span> <span class="n">partial_charges_str</span><span class="p">)</span>

        <span class="c1"># If the file format is &quot;pdb&quot; using OEWriteMolecule() rearranges the atoms (hydrogens are pushed to the bottom)</span>
        <span class="c1"># Issue #475 (https://github.com/openforcefield/openff-toolkit/issues/475)</span>
        <span class="c1"># dfhahn&#39;s workaround: Using OEWritePDBFile does not alter the atom arrangement</span>
        <span class="k">if</span> <span class="n">file_format</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;pdb&quot;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NumConfs</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">conf</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetConfs</span><span class="p">():</span>
                    <span class="n">oechem</span><span class="o">.</span><span class="n">OEWritePDBFile</span><span class="p">(</span><span class="n">ofs</span><span class="p">,</span> <span class="n">conf</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEOFlavor_PDB_BONDS</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">oechem</span><span class="o">.</span><span class="n">OEWritePDBFile</span><span class="p">(</span><span class="n">ofs</span><span class="p">,</span> <span class="n">oemol</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEOFlavor_PDB_BONDS</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEWriteMolecule</span><span class="p">(</span><span class="n">ofs</span><span class="p">,</span> <span class="n">oemol</span><span class="p">)</span>
        <span class="n">ofs</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_turn_oemolbase_sd_charges_into_partial_charges</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Process an OEMolBase object and check to see whether it has an SD data pair</span>
<span class="sd">        where the tag is &quot;atom.dprop.PartialCharge&quot;, indicating that it has a list of</span>
<span class="sd">        atomic partial charges. If so, apply those charges to the OEAtoms in the OEMolBase,</span>
<span class="sd">        and delete the SD data pair.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        oemol</span>
<span class="sd">            The molecule to process</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        charges_are_present</span>
<span class="sd">            Whether charges are present in the SD file. This is necessary because OEAtoms</span>
<span class="sd">            have a default partial charge of 0.0, which makes truly zero-charge molecules</span>
<span class="sd">            (eg &quot;N2&quot;, &quot;Ar&quot;...) indistinguishable from molecules for which partial charges</span>
<span class="sd">            have not been assigned. The OFF Toolkit allows this distinction with</span>
<span class="sd">            mol.partial_charges=None. In order to complete roundtrips within the OFFMol</span>
<span class="sd">            spec, we must interpret the presence or absence of this tag as a proxy for</span>
<span class="sd">            mol.partial_charges=None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGetSDDataPairs</span><span class="p">(</span><span class="n">oemol</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">dp</span><span class="o">.</span><span class="n">GetTag</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;atom.dprop.PartialCharge&quot;</span><span class="p">:</span>
                <span class="n">charges_str</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGetSDData</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="s2">&quot;atom.dprop.PartialCharge&quot;</span><span class="p">)</span>
                <span class="n">charges_unitless</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">charges_str</span><span class="o">.</span><span class="n">split</span><span class="p">()]</span>
                <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">charges_unitless</span><span class="p">)</span> <span class="o">==</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NumAtoms</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">charge</span><span class="p">,</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">charges_unitless</span><span class="p">,</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()):</span>
                    <span class="n">oeatom</span><span class="o">.</span><span class="n">SetPartialCharge</span><span class="p">(</span><span class="n">charge</span><span class="p">)</span>
                <span class="n">oechem</span><span class="o">.</span><span class="n">OEDeleteSDData</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="s2">&quot;atom.dprop.PartialCharge&quot;</span><span class="p">)</span>
                <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">_read_oemolistream_molecules</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">oemolistream</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
        <span class="n">file_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Reads and return the Molecules in a OEMol input stream.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        oemolistream</span>
<span class="sd">            The OEMol input stream to read from.</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            If false, raises an exception if oemol contains undefined stereochemistry.</span>
<span class="sd">        file_path</span>
<span class="sd">            The path to the mol2 file. This is used exclusively to make</span>
<span class="sd">            the error message more meaningful when the mol2 files doesn&#39;t</span>
<span class="sd">            use Tripos atom types.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecules</span>
<span class="sd">            The list of Molecule objects in the stream.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">mols</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">()</span>
        <span class="k">while</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEReadMolecule</span><span class="p">(</span><span class="n">oemolistream</span><span class="p">,</span> <span class="n">oemol</span><span class="p">):</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEPerceiveChiral</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEAssignAromaticFlags</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAroModel_MDL</span><span class="p">)</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OE3DToInternalStereo</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

            <span class="c1"># If this is either a multi-conformer or multi-molecule SD file, check to see if there are partial charges</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">oemolistream</span><span class="o">.</span><span class="n">GetFormat</span><span class="p">()</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEFormat_SDF</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">hasattr</span><span class="p">(</span>
                <span class="n">oemol</span><span class="p">,</span> <span class="s2">&quot;GetConfs&quot;</span>
            <span class="p">):</span>
                <span class="c1"># The openFF toolkit treats each conformer in a &quot;multiconformer&quot; SDF as</span>
                <span class="c1"># a separate molecule.</span>
                <span class="c1"># https://github.com/openforcefield/openff-toolkit/issues/202</span>
                <span class="c1"># Note that there is ambiguity about how SD data and &quot;multiconformer&quot; SD files should be stored.</span>
                <span class="c1"># As a result, we have to do some weird stuff below, as discussed in</span>
                <span class="c1"># https://docs.eyesopen.com/toolkits/python/oechemtk/oemol.html#dude-where-s-my-sd-data</span>

                <span class="c1"># Jeff: I was unable to find a way to distinguish whether a SDF was multiconformer or not.</span>
                <span class="c1"># The logic below should handle either single- or multi-conformer SDFs.</span>
                <span class="k">for</span> <span class="n">conf</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetConfIter</span><span class="p">():</span>
                    <span class="c1"># First, we turn &quot;conf&quot; into an OEMCMol (OE multiconformer mol), since OTHER file formats</span>
                    <span class="c1"># really are multiconformer, and we will eventually feed this into the `from_openeye` function,</span>
                    <span class="c1"># which is made to ingest multiconformer mols.</span>
                    <span class="n">this_conf_oemcmol</span> <span class="o">=</span> <span class="n">conf</span><span class="o">.</span><span class="n">GetMCMol</span><span class="p">()</span>

                    <span class="c1"># Then, we take any SD data pairs that were on the oemol, and copy them on to &quot;this_conf_oemcmol&quot;.</span>
                    <span class="c1"># These SD pairs will be populated if we&#39;re dealing with a single-conformer SDF.</span>
                    <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGetSDDataPairs</span><span class="p">(</span><span class="n">oemol</span><span class="p">):</span>
                        <span class="n">oechem</span><span class="o">.</span><span class="n">OESetSDData</span><span class="p">(</span>
                            <span class="n">this_conf_oemcmol</span><span class="p">,</span> <span class="n">dp</span><span class="o">.</span><span class="n">GetTag</span><span class="p">(),</span> <span class="n">dp</span><span class="o">.</span><span class="n">GetValue</span><span class="p">()</span>
                        <span class="p">)</span>
                    <span class="c1"># On the other hand, these SD pairs will be populated if we&#39;re dealing with a MULTI-conformer SDF.</span>
                    <span class="k">for</span> <span class="n">dp</span> <span class="ow">in</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGetSDDataPairs</span><span class="p">(</span><span class="n">conf</span><span class="p">):</span>
                        <span class="n">oechem</span><span class="o">.</span><span class="n">OESetSDData</span><span class="p">(</span>
                            <span class="n">this_conf_oemcmol</span><span class="p">,</span> <span class="n">dp</span><span class="o">.</span><span class="n">GetTag</span><span class="p">(),</span> <span class="n">dp</span><span class="o">.</span><span class="n">GetValue</span><span class="p">()</span>
                        <span class="p">)</span>
                    <span class="c1"># This function fishes out the special SD data tag we use for partial charge</span>
                    <span class="c1"># (&quot;atom.dprop.PartialCharge&quot;), and applies those as OETK-supported partial charges on the OEAtoms</span>
                    <span class="n">has_charges</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_turn_oemolbase_sd_charges_into_partial_charges</span><span class="p">(</span>
                        <span class="n">this_conf_oemcmol</span>
                    <span class="p">)</span>

                    <span class="c1"># Finally, we feed the molecule into `from_openeye`, where it converted into an OFFMol</span>
                    <span class="n">mol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
                        <span class="n">this_conf_oemcmol</span><span class="p">,</span>
                        <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="n">allow_undefined_stereo</span><span class="p">,</span>
                        <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span><span class="p">,</span>
                    <span class="p">)</span>

                    <span class="c1"># If the molecule didn&#39;t even have the `PartialCharges` tag, we set it from zeroes to None here.</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">has_charges</span><span class="p">):</span>
                        <span class="n">mol</span><span class="o">.</span><span class="n">partial_charges</span> <span class="o">=</span> <span class="kc">None</span>
                    <span class="n">mols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="c1"># In case this is being read from a SINGLE-molecule SD file, convert the SD field where we</span>
                <span class="c1"># stash partial charges into actual per-atom partial charges</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_turn_oemolbase_sd_charges_into_partial_charges</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
                <span class="n">mol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
                    <span class="n">oemol</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="n">allow_undefined_stereo</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span>
                <span class="p">)</span>
                <span class="n">mols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>

            <span class="c1"># Check if this is an AMBER-produced mol2 file, which we can not load because they use GAFF atom types.</span>
            <span class="k">if</span> <span class="n">oemolistream</span><span class="o">.</span><span class="n">GetFormat</span><span class="p">()</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEFormat_MOL2</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_check_mol2_gaff_atom_type</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">mols</span>

    <span class="k">def</span> <span class="nf">_smarts_to_networkx</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">substructure_smarts</span><span class="p">):</span>
        <span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">qmol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEQMol</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEParseSmiles</span><span class="p">(</span><span class="n">qmol</span><span class="p">,</span> <span class="n">substructure_smarts</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">SMILESParseError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error parsing SMARTS &#39;</span><span class="si">{</span><span class="n">substructure_smarts</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OEAssignHybridization</span><span class="p">(</span><span class="n">qmol</span><span class="p">)</span>

        <span class="n">graph</span><span class="p">:</span> <span class="n">nx</span><span class="o">.</span><span class="n">classes</span><span class="o">.</span><span class="n">graph</span><span class="o">.</span><span class="n">Graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">qmol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="n">atomic_number</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()</span>

            <span class="n">graph</span><span class="o">.</span><span class="n">add_node</span><span class="p">(</span>
                <span class="n">atom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">(),</span>
                <span class="n">atomic_number</span><span class="o">=</span><span class="n">atomic_number</span><span class="p">,</span>
                <span class="n">formal_charge</span><span class="o">=</span><span class="n">atom</span><span class="o">.</span><span class="n">GetFormalCharge</span><span class="p">(),</span>
                <span class="n">map_index</span><span class="o">=</span><span class="n">atom</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">(),</span>
            <span class="p">)</span>
        <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">qmol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
            <span class="n">bond_order</span> <span class="o">=</span> <span class="n">bond</span><span class="o">.</span><span class="n">GetOrder</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">bond_order</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">SMILESParseError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;A bond in &#39;</span><span class="si">{</span><span class="n">substructure_smarts</span><span class="si">}</span><span class="s2"> has order 0&quot;</span><span class="p">)</span>

            <span class="n">graph</span><span class="o">.</span><span class="n">add_edge</span><span class="p">(</span>
                <span class="n">bond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">(),</span>
                <span class="n">bond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">(),</span>
                <span class="n">bond_order</span><span class="o">=</span><span class="n">bond_order</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">graph</span>

    <span class="k">def</span> <span class="nf">_assign_aromaticity_and_stereo_from_3d</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">offmol</span><span class="p">):</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">offmol</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">()</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEPerceiveChiral</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OE3DToInternalStereo</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="c1"># Aromaticity is re-perceived in this call</span>
        <span class="n">offmol_w_stereo_and_aro</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">offmol_w_stereo_and_aro</span>

    <span class="k">def</span> <span class="nf">enumerate_protomers</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">,</span>
        <span class="n">max_states</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Enumerate the formal charges of a molecule to generate different protomers.</span>
<span class="sd">        Note that, in cases where the input molecule has an uncommon protonation state</span>
<span class="sd">        (for example ``[NH2-]``), the input molecule may not be included in the output.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule whose state we should enumerate</span>

<span class="sd">        max_states</span>
<span class="sd">            The maximum number of protomer states to be returned. If 0, the default,</span>
<span class="sd">            attempt to return all protomers.</span>

<span class="sd">        Returns</span>
<span class="sd">        ------</span>
<span class="sd">        molecules</span>
<span class="sd">            A list of the protomers of the input molecules, including the input molecule if found</span>
<span class="sd">            by Quacpac and not pruned by `max_states`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oequacpac</span>

        <span class="n">options</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEFormalChargeOptions</span><span class="p">()</span>
        <span class="n">options</span><span class="o">.</span><span class="n">SetMaxCount</span><span class="p">(</span><span class="n">max_states</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">[</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
                <span class="n">protomer</span><span class="p">,</span>
                <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">protomer</span> <span class="ow">in</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEEnumerateFormalCharges</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="o">=</span><span class="n">molecule</span><span class="p">),</span>
                <span class="n">options</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="p">]</span>

    <span class="k">def</span> <span class="nf">enumerate_stereoisomers</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">,</span>
        <span class="n">undefined_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">max_isomers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span>
        <span class="n">rationalise</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Enumerate the stereocenters and bonds of the current molecule.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule whose state we should enumerate</span>

<span class="sd">        undefined_only</span>
<span class="sd">            If we should enumerate all stereocenters and bonds or only those with undefined stereochemistry</span>

<span class="sd">        max_isomers</span>
<span class="sd">            The maximum amount of molecules that should be returned</span>

<span class="sd">        rationalise</span>
<span class="sd">            If we should try to build and rationalise the molecule to ensure it can exist</span>


<span class="sd">        Returns</span>
<span class="sd">        --------</span>
<span class="sd">        molecules</span>
<span class="sd">            A list of openff.toolkit.topology.Molecule instances</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span><span class="p">,</span> <span class="n">oeomega</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="o">=</span><span class="n">molecule</span><span class="p">)</span>

        <span class="c1"># arguments for this function can be found here</span>
        <span class="c1"># &lt;https://docs.eyesopen.com/toolkits/python/omegatk/OEConfGenFunctions/OEFlipper.html?highlight=stereoisomers&gt;</span>

        <span class="n">molecules</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">isomer</span> <span class="ow">in</span> <span class="n">oeomega</span><span class="o">.</span><span class="n">OEFlipper</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="ow">not</span> <span class="n">undefined_only</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">rationalise</span><span class="p">:</span>
                <span class="c1"># try and determine if the molecule is reasonable by generating a conformer with</span>
                <span class="c1"># strict stereo, like embedding in rdkit</span>
                <span class="n">omega</span> <span class="o">=</span> <span class="n">oeomega</span><span class="o">.</span><span class="n">OEOmega</span><span class="p">()</span>
                <span class="n">omega</span><span class="o">.</span><span class="n">SetMaxConfs</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">omega</span><span class="o">.</span><span class="n">SetCanonOrder</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
                <span class="c1"># Don&#39;t generate random stereoisomer if not specified</span>
                <span class="n">omega</span><span class="o">.</span><span class="n">SetStrictStereo</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
                <span class="n">mol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">(</span><span class="n">isomer</span><span class="p">)</span>
                <span class="n">status</span> <span class="o">=</span> <span class="n">omega</span><span class="p">(</span><span class="n">mol</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">status</span><span class="p">:</span>
                    <span class="n">isomol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">isomol</span> <span class="o">!=</span> <span class="n">molecule</span><span class="p">:</span>
                        <span class="n">molecules</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">isomol</span><span class="p">)</span>

            <span class="k">else</span><span class="p">:</span>
                <span class="n">isomol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span><span class="n">isomer</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">isomol</span> <span class="o">!=</span> <span class="n">molecule</span><span class="p">:</span>
                    <span class="n">molecules</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">isomol</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">molecules</span><span class="p">[:</span><span class="n">max_isomers</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">enumerate_tautomers</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">,</span> <span class="n">max_states</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">20</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Enumerate the possible tautomers of the current molecule</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule whose state we should enumerate</span>

<span class="sd">        max_states</span>
<span class="sd">            The maximum amount of molecules that should be returned</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecules</span>
<span class="sd">            A list of openff.toolkit.topology.Molecule instances excluding the input molecule.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oequacpac</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="o">=</span><span class="n">molecule</span><span class="p">)</span>

        <span class="n">tautomers</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c1"># set the options</span>
        <span class="n">tautomer_options</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OETautomerOptions</span><span class="p">()</span>
        <span class="n">tautomer_options</span><span class="o">.</span><span class="n">SetApplyWarts</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">tautomer_options</span><span class="o">.</span><span class="n">SetMaxTautomersGenerated</span><span class="p">(</span><span class="n">max_states</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">tautomer_options</span><span class="o">.</span><span class="n">SetSaveStereo</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="c1"># this aligns the outputs of rdkit and openeye for the example cases</span>
        <span class="n">tautomer_options</span><span class="o">.</span><span class="n">SetCarbonHybridization</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">tautomer</span> <span class="ow">in</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEEnumerateTautomers</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">tautomer_options</span><span class="p">):</span>
            <span class="c1"># remove the input tautomer from the output</span>
            <span class="n">taut</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
                <span class="n">tautomer</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">taut</span> <span class="o">!=</span> <span class="n">molecule</span><span class="p">:</span>
                <span class="n">tautomers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
                        <span class="n">tautomer</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span>
                    <span class="p">)</span>
                <span class="p">)</span>

        <span class="k">return</span> <span class="n">tautomers</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_check_mol2_gaff_atom_type</span><span class="p">(</span><span class="n">molecule</span><span class="p">,</span> <span class="n">file_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Attempts to detect the presence of GAFF atom types in a molecule loaded from a mol2 file.</span>

<span class="sd">        For now, this raises a ``GAFFAtomTypeWarning`` if the molecule</span>
<span class="sd">        include Osmium and Holmium atoms, which have GAFF types OS and</span>
<span class="sd">        HO respectively.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The loaded molecule.</span>
<span class="sd">        file_path</span>
<span class="sd">            The path to the mol2 file. This is used exclusively to make</span>
<span class="sd">            the error message more meaningful.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Handle default.</span>
        <span class="k">if</span> <span class="n">file_path</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Append a &#39;:&#39; character that will separate the file</span>
            <span class="c1"># path from the molecule string representation.</span>
            <span class="n">file_path</span> <span class="o">=</span> <span class="n">file_path</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span>
        <span class="c1"># atomic_number: (GAFF_type, element_name)</span>
        <span class="n">warning_atomic_numbers</span> <span class="o">=</span> <span class="p">{</span><span class="mi">76</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;OS&quot;</span><span class="p">,</span> <span class="s2">&quot;Osmium&quot;</span><span class="p">),</span> <span class="mi">67</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;HO&quot;</span><span class="p">,</span> <span class="s2">&quot;Holmium&quot;</span><span class="p">)}</span>

        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molecule</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">atom_type</span><span class="p">,</span> <span class="n">element_name</span> <span class="o">=</span> <span class="n">warning_atomic_numbers</span><span class="p">[</span><span class="n">atom</span><span class="o">.</span><span class="n">atomic_number</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">warn_msg</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s1">&#39;OpenEye interpreted the type &quot;</span><span class="si">{</span><span class="n">atom_type</span><span class="si">}</span><span class="s1">&quot; in </span><span class="si">{</span><span class="n">file_path</span><span class="si">}{</span><span class="n">molecule</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s1">&#39;</span>
                    <span class="sa">f</span><span class="s2">&quot; as </span><span class="si">{</span><span class="n">element_name</span><span class="si">}</span><span class="s2">. Does your mol2 file uses Tripos SYBYL atom types?&quot;</span>
                    <span class="s2">&quot; Other atom types such as GAFF are not supported.&quot;</span>
                <span class="p">)</span>
                <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="n">warn_msg</span><span class="p">,</span> <span class="n">GAFFAtomTypeWarning</span><span class="p">,</span> <span class="n">stacklevel</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_openeye_cip_atom_stereochemistry</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oeatom</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine CIP stereochemistry (R/S) for the specified atom</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        oemol</span>
<span class="sd">            The molecule of interest</span>
<span class="sd">        oeatom</span>
<span class="sd">            The atom whose stereochemistry is to be computed</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        stereochemistry</span>
<span class="sd">            &#39;R&#39;, &#39;S&#39;, or None if no stereochemistry is specified or the atom is not a stereocenter</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">HasStereoSpecified</span><span class="p">():</span>
            <span class="c1"># No stereochemical information has been stored, so this could be unknown stereochemistry</span>
            <span class="c1"># TODO: Should we raise an exception?</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">cip</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEPerceiveCIPStereo</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oeatom</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">cip</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECIPAtomStereo_S</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;S&quot;</span>
        <span class="k">elif</span> <span class="n">cip</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECIPAtomStereo_R</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;R&quot;</span>
        <span class="k">elif</span> <span class="n">cip</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECIPAtomStereo_NotStereo</span><span class="p">:</span>
            <span class="c1"># Not a stereocenter</span>
            <span class="c1"># TODO: Should this be a different case from ``None``?</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_openeye_cip_bond_stereochemistry</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oebond</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Determine CIP stereochemistry (E/Z) for the specified bond</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        oemol</span>
<span class="sd">            The molecule of interest</span>
<span class="sd">        oebond</span>
<span class="sd">            The bond whose stereochemistry is to be computed</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        stereochemistry</span>
<span class="sd">            &#39;E&#39;, &#39;Z&#39;, or None if stereochemistry is unspecified or the bond is not a stereo bond</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">oebond</span><span class="o">.</span><span class="n">HasStereoSpecified</span><span class="p">():</span>
            <span class="c1"># No stereochemical information has been stored, so this could be unknown stereochemistry</span>
            <span class="c1"># TODO: Should we raise an exception?</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="n">cip</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEPerceiveCIPStereo</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oebond</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">cip</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECIPBondStereo_E</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;E&quot;</span>
        <span class="k">elif</span> <span class="n">cip</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECIPBondStereo_Z</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;Z&quot;</span>
        <span class="k">elif</span> <span class="n">cip</span> <span class="o">==</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECIPBondStereo_NotStereo</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">from_openeye</span><span class="p">(</span>
        <span class="n">oemol</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a Molecule from an OpenEye molecule. If the OpenEye molecule has</span>
<span class="sd">        implicit hydrogens, this function will make them explicit.</span>

<span class="sd">        ``OEAtom`` s have a different set of allowed value for partial charges than</span>
<span class="sd">        ``openff.toolkit.topology.Molecule`` s. In the OpenEye toolkits, partial charges</span>
<span class="sd">        are stored on individual ``OEAtom`` s, and their values are initialized to ``0.0``.</span>
<span class="sd">        In the Open Force Field Toolkit, an ``openff.toolkit.topology.Molecule``&#39;s</span>
<span class="sd">        ``partial_charges`` attribute is initialized to ``None`` and can be set to a</span>
<span class="sd">        unit-wrapped numpy array with units of</span>
<span class="sd">        elementary charge. The Open Force</span>
<span class="sd">        Field Toolkit considers an ``OEMol`` where every ``OEAtom`` has a partial</span>
<span class="sd">        charge of ``float(&#39;nan&#39;)`` to be equivalent to an Open Force Field Toolkit `Molecule`&#39;s</span>
<span class="sd">        ``partial_charges = None``.</span>
<span class="sd">        This assumption is made in both ``to_openeye`` and ``from_openeye``.</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        oemol</span>
<span class="sd">            An OpenEye molecule</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            If false, raises an exception if oemol contains undefined stereochemistry.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecule</span>
<span class="sd">            An OpenFF molecule</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        Create a Molecule from an OpenEye OEMol</span>

<span class="sd">        &gt;&gt;&gt; from openeye import oechem</span>
<span class="sd">        &gt;&gt;&gt; from openff.toolkit._tests.utils import get_data_file_path</span>
<span class="sd">        &gt;&gt;&gt; ifs = oechem.oemolistream(get_data_file_path(&#39;systems/monomers/ethanol.mol2&#39;))</span>
<span class="sd">        &gt;&gt;&gt; oemols = list(ifs.GetOEGraphMols())</span>

<span class="sd">        &gt;&gt;&gt; toolkit_wrapper = OpenEyeToolkitWrapper()</span>
<span class="sd">        &gt;&gt;&gt; molecule = toolkit_wrapper.from_openeye(oemols[0])</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span> <span class="nn">math</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="c1"># Save the existing SD tags, if any, since they&#39;re lost in the cast to</span>
        <span class="c1"># OEMol of the input is OEGraphMol. See issue #1711</span>
        <span class="n">existing_sd_tags</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">pair</span><span class="o">.</span><span class="n">GetTag</span><span class="p">():</span> <span class="n">pair</span><span class="o">.</span><span class="n">GetValue</span><span class="p">()</span> <span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGetSDDataIter</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="p">}</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="n">components</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEDetermineComponents</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">components</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">warnings</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;OpenEye OEMol passed to from_openeye consists of more than one molecule, consider running &quot;</span>
                          <span class="s2">&quot;something like https://docs.eyesopen.com/toolkits/python/oechemtk/oechem_examples/&quot;</span>
                          <span class="s2">&quot;oechem_example_parts2mols.html or splitting input SMILES at &#39;.&#39;s to get separate molecules &quot;</span>
                          <span class="s2">&quot;and pass them to from_openeye one at a time. While this is supported for &quot;</span>
                          <span class="s2">&quot;legacy reasons, OpenFF Molecule objects are not supposed to contain disconnected chemical &quot;</span>
                          <span class="s2">&quot;graphs and this may result in undefined behavior later on. The OpenFF ecosystem is built &quot;</span>
                          <span class="s2">&quot;to handle multiple molecules, but they should be in a Topology object, ex: &quot;</span>
                          <span class="s2">&quot;top = Topology.from_molecules([mol1, mol2])&quot;</span><span class="p">,</span>
                          <span class="n">MultipleComponentsInMoleculeWarning</span><span class="p">,</span>
                          <span class="n">stacklevel</span><span class="o">=</span><span class="mi">2</span>
                          <span class="p">)</span>

        <span class="c1"># Add explicit hydrogens if they&#39;re implicit</span>
        <span class="k">if</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEHasImplicitHydrogens</span><span class="p">(</span><span class="n">oemol</span><span class="p">):</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEAddExplicitHydrogens</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="c1"># TODO: Is there any risk to perceiving aromaticity here instead of later?</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEAssignAromaticFlags</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAroModel_MDL</span><span class="p">)</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OEPerceiveChiral</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="c1"># Check that all stereo is specified</span>
        <span class="c1"># Potentially better OE stereo check: OEFlipper  Toolkits - - Python</span>
        <span class="c1"># https: // docs.eyesopen.com / toolkits / python / omegatk / OEConfGenFunctions / OEFlipper.html</span>

        <span class="n">unspec_chiral</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">unspec_db</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">problematic_atoms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="n">problematic_bonds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">IsChiral</span><span class="p">():</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">oeatom</span><span class="o">.</span><span class="n">HasStereoSpecified</span><span class="p">()):</span>
                    <span class="n">unspec_chiral</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">problematic_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">oebond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">oebond</span><span class="o">.</span><span class="n">IsChiral</span><span class="p">():</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">oebond</span><span class="o">.</span><span class="n">HasStereoSpecified</span><span class="p">()):</span>
                    <span class="n">unspec_db</span> <span class="o">=</span> <span class="kc">True</span>
                    <span class="n">problematic_bonds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oebond</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">unspec_chiral</span> <span class="ow">or</span> <span class="n">unspec_db</span><span class="p">:</span>

            <span class="k">def</span> <span class="nf">oeatom_to_str</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
                <span class="k">return</span> <span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;atomic num: </span><span class="si">{</span><span class="n">oeatom</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()</span><span class="si">}</span><span class="s2">, &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;name: </span><span class="si">{</span><span class="n">oeatom</span><span class="o">.</span><span class="n">GetName</span><span class="p">()</span><span class="si">}</span><span class="s2">, &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;idx: </span><span class="si">{</span><span class="n">oeatom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span><span class="si">}</span><span class="s2">, &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;aromatic: </span><span class="si">{</span><span class="n">oeatom</span><span class="o">.</span><span class="n">IsAromatic</span><span class="p">()</span><span class="si">}</span><span class="s2">, &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;chiral: </span><span class="si">{</span><span class="n">oeatom</span><span class="o">.</span><span class="n">IsChiral</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>


            <span class="k">def</span> <span class="nf">oebond_to_str</span><span class="p">(</span><span class="n">oebond</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
                <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;order: </span><span class="si">{</span><span class="n">oebond</span><span class="o">.</span><span class="n">GetOrder</span><span class="p">()</span><span class="si">}</span><span class="s2">, chiral: </span><span class="si">{</span><span class="n">oebond</span><span class="o">.</span><span class="n">IsChiral</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span>

            <span class="k">def</span> <span class="nf">describe_oeatom</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
                <span class="n">description</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;Atom </span><span class="si">{</span><span class="n">oeatom_to_str</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span><span class="si">}</span><span class="s2"> with bonds:&quot;</span>
                <span class="k">for</span> <span class="n">oebond</span> <span class="ow">in</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
                    <span class="n">description</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">bond </span><span class="si">{</span><span class="n">oebond_to_str</span><span class="p">(</span><span class="n">oebond</span><span class="p">)</span><span class="si">}</span><span class="s2"> to atom </span><span class="si">{</span><span class="n">oeatom_to_str</span><span class="p">(</span><span class="n">oebond</span><span class="o">.</span><span class="n">GetNbr</span><span class="p">(</span><span class="n">oeatom</span><span class="p">))</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">return</span> <span class="n">description</span>

            <span class="k">if</span> <span class="p">(</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">problematic_atoms</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">problematic_bonds</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span>
            <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">allow_undefined_stereo</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;OEMol has unspecified stereochemistry. </span><span class="si">{</span><span class="n">oemol</span><span class="o">.</span><span class="n">GetTitle</span><span class="p">()</span><span class="si">=}</span><span class="s2">&quot;</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">problematic_atoms</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">+=</span> <span class="s2">&quot;Problematic atoms are:</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="k">for</span> <span class="n">problematic_atom</span> <span class="ow">in</span> <span class="n">problematic_atoms</span><span class="p">:</span>
                        <span class="n">msg</span> <span class="o">+=</span> <span class="n">describe_oeatom</span><span class="p">(</span><span class="n">problematic_atom</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">problematic_bonds</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">msg</span> <span class="o">+=</span> <span class="sa">f</span><span class="s2">&quot;Problematic bonds are: </span><span class="si">{</span><span class="n">problematic_bonds</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>

                <span class="k">raise</span> <span class="n">UndefinedStereochemistryError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Unable to make OFFMol from OEMol: </span><span class="si">{</span><span class="n">msg</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="n">_cls</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="kn">from</span> <span class="nn">openff.toolkit.topology.molecule</span> <span class="kn">import</span> <span class="n">Molecule</span>

            <span class="n">_cls</span> <span class="o">=</span> <span class="n">Molecule</span>

        <span class="n">molecule</span> <span class="o">=</span> <span class="n">_cls</span><span class="p">()</span>

        <span class="c1"># OEMol.GetTitle() happens to default to an empty string</span>
        <span class="k">if</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetTitle</span><span class="p">()</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetTitle</span><span class="p">()</span>

        <span class="c1"># Attach any SD tag information we saved before casting input to new OEMol</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">existing_sd_tags</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">_properties</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>

        <span class="n">off_to_oe_idx</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>  <span class="c1"># {oemol_idx: molecule_idx}</span>
        <span class="n">atom_mapping</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="n">oe_idx</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>
            <span class="n">map_id</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">()</span>
            <span class="n">atomic_number</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()</span>
            <span class="c1"># Carry with implicit units of elementary charge for faster route through _add_atom</span>
            <span class="n">formal_charge</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetFormalCharge</span><span class="p">()</span>
            <span class="n">explicit_valence</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetExplicitValence</span><span class="p">()</span>
            <span class="c1"># Implicit hydrogens are never added to D- and F- block elements,</span>
            <span class="c1"># and the MDL valence is always the explicit valence for these</span>
            <span class="c1"># elements, so this does not count radical electrons in these blocks.</span>
            <span class="n">mdl_valence</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMDLGetValence</span><span class="p">(</span>
                <span class="n">atomic_number</span><span class="p">,</span> <span class="n">formal_charge</span><span class="p">,</span> <span class="n">explicit_valence</span>
            <span class="p">)</span>
            <span class="n">number_radical_electrons</span> <span class="o">=</span> <span class="n">mdl_valence</span> <span class="o">-</span> <span class="p">(</span>
                <span class="n">oeatom</span><span class="o">.</span><span class="n">GetImplicitHCount</span><span class="p">()</span> <span class="o">+</span> <span class="n">explicit_valence</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">number_radical_electrons</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">RadicalsNotSupportedError</span><span class="p">(</span>
                    <span class="s2">&quot;The OpenFF Toolkit does not currently support parsing molecules with radicals. &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;Found </span><span class="si">{</span><span class="n">number_radical_electrons</span><span class="si">}</span><span class="s2"> radical electrons on molecule &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">oechem</span><span class="o">.</span><span class="n">OECreateSmiString</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span><span class="si">}</span><span class="s2">.&quot;</span>
                <span class="p">)</span>

            <span class="n">is_aromatic</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">IsAromatic</span><span class="p">()</span>
            <span class="n">stereochemistry</span> <span class="o">=</span> <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_openeye_cip_atom_stereochemistry</span><span class="p">(</span>
                <span class="n">oemol</span><span class="p">,</span> <span class="n">oeatom</span>
            <span class="p">)</span>
            <span class="c1"># stereochemistry = self._openeye_cip_atom_stereochemistry(oemol, oeatom)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetName</span><span class="p">()</span>

            <span class="c1"># Transfer in hierarchy metadata</span>
            <span class="n">metadata_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEHasResidues</span><span class="p">(</span><span class="n">oemol</span><span class="p">):</span>
                <span class="n">metadata_dict</span><span class="p">[</span><span class="s2">&quot;residue_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomGetResidue</span><span class="p">(</span>
                    <span class="n">oeatom</span>
                <span class="p">)</span><span class="o">.</span><span class="n">GetName</span><span class="p">()</span>
                <span class="n">metadata_dict</span><span class="p">[</span><span class="s2">&quot;residue_number&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomGetResidue</span><span class="p">(</span>
                    <span class="n">oeatom</span>
                <span class="p">)</span><span class="o">.</span><span class="n">GetResidueNumber</span><span class="p">()</span>
                <span class="n">metadata_dict</span><span class="p">[</span><span class="s2">&quot;insertion_code&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomGetResidue</span><span class="p">(</span>
                    <span class="n">oeatom</span>
                <span class="p">)</span><span class="o">.</span><span class="n">GetInsertCode</span><span class="p">()</span>
                <span class="n">metadata_dict</span><span class="p">[</span><span class="s2">&quot;chain_id&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomGetResidue</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span><span class="o">.</span><span class="n">GetChainID</span><span class="p">()</span>
            <span class="c1"># print(&#39;from&#39;, metadata_dict)</span>

            <span class="n">atom_index</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">_add_atom</span><span class="p">(</span>
                <span class="n">atomic_number</span><span class="p">,</span>
                <span class="n">formal_charge</span><span class="p">,</span>
                <span class="n">is_aromatic</span><span class="p">,</span>
                <span class="n">stereochemistry</span><span class="o">=</span><span class="n">stereochemistry</span><span class="p">,</span>
                <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span>
                <span class="n">metadata</span><span class="o">=</span><span class="n">metadata_dict</span><span class="p">,</span>
                <span class="n">invalidate_cache</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">atom_index</span>  <span class="c1"># store for mapping oeatom to molecule atom indices below</span>
            <span class="p">)</span>
            <span class="n">atom_mapping</span><span class="p">[</span><span class="n">atom_index</span><span class="p">]</span> <span class="o">=</span> <span class="n">map_id</span>

        <span class="n">molecule</span><span class="o">.</span><span class="n">_invalidate_cached_properties</span><span class="p">()</span>

        <span class="c1"># If we have a full / partial atom map add it to the molecule. Zeroes 0</span>
        <span class="c1"># indicates no mapping</span>
        <span class="k">if</span> <span class="p">{</span><span class="o">*</span><span class="n">atom_mapping</span><span class="o">.</span><span class="n">values</span><span class="p">()}</span> <span class="o">!=</span> <span class="p">{</span><span class="mi">0</span><span class="p">}:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">_properties</span><span class="p">[</span><span class="s2">&quot;atom_map&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">idx</span><span class="p">:</span> <span class="n">map_idx</span> <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">map_idx</span> <span class="ow">in</span> <span class="n">atom_mapping</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="k">if</span> <span class="n">map_idx</span> <span class="o">!=</span> <span class="mi">0</span>
            <span class="p">}</span>

        <span class="k">for</span> <span class="n">oebond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
            <span class="n">atom1_index</span> <span class="o">=</span> <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">oebond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">()]</span>
            <span class="n">atom2_index</span> <span class="o">=</span> <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">oebond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">()]</span>
            <span class="n">bond_order</span> <span class="o">=</span> <span class="n">oebond</span><span class="o">.</span><span class="n">GetOrder</span><span class="p">()</span>
            <span class="n">is_aromatic</span> <span class="o">=</span> <span class="n">oebond</span><span class="o">.</span><span class="n">IsAromatic</span><span class="p">()</span>
            <span class="n">stereochemistry</span> <span class="o">=</span> <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_openeye_cip_bond_stereochemistry</span><span class="p">(</span>
                <span class="n">oemol</span><span class="p">,</span> <span class="n">oebond</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">oebond</span><span class="o">.</span><span class="n">HasData</span><span class="p">(</span><span class="s2">&quot;fractional_bond_order&quot;</span><span class="p">):</span>
                <span class="n">fractional_bond_order</span> <span class="o">=</span> <span class="n">oebond</span><span class="o">.</span><span class="n">GetData</span><span class="p">(</span><span class="s2">&quot;fractional_bond_order&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">fractional_bond_order</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="n">molecule</span><span class="o">.</span><span class="n">_add_bond</span><span class="p">(</span>
                <span class="n">atom1_index</span><span class="p">,</span>
                <span class="n">atom2_index</span><span class="p">,</span>
                <span class="n">bond_order</span><span class="p">,</span>
                <span class="n">is_aromatic</span><span class="o">=</span><span class="n">is_aromatic</span><span class="p">,</span>
                <span class="n">stereochemistry</span><span class="o">=</span><span class="n">stereochemistry</span><span class="p">,</span>
                <span class="n">fractional_bond_order</span><span class="o">=</span><span class="n">fractional_bond_order</span><span class="p">,</span>
                <span class="n">invalidate_cache</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="n">molecule</span><span class="o">.</span><span class="n">_invalidate_cached_properties</span><span class="p">()</span>

        <span class="c1"># TODO: Copy conformations, if present</span>
        <span class="c1"># TODO: Come up with some scheme to know when to import coordinates</span>
        <span class="c1"># From SMILES: no</span>
        <span class="c1"># From MOL2: maybe</span>
        <span class="c1"># From other: maybe</span>
        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="s2">&quot;GetConfs&quot;</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">conf</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetConfs</span><span class="p">():</span>
                <span class="n">n_atoms</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">n_atoms</span>
                <span class="c1"># Store with implicit units until we&#39;re sure this conformer exists</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="p">[</span><span class="n">n_atoms</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">oe_id</span> <span class="ow">in</span> <span class="n">conf</span><span class="o">.</span><span class="n">GetCoords</span><span class="p">()</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="c1"># implicitly in angstrom</span>
                    <span class="n">off_atom_coords</span> <span class="o">=</span> <span class="n">conf</span><span class="o">.</span><span class="n">GetCoords</span><span class="p">()[</span><span class="n">oe_id</span><span class="p">]</span>
                    <span class="n">off_atom_index</span> <span class="o">=</span> <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">oe_id</span><span class="p">]</span>
                    <span class="n">positions</span><span class="p">[</span><span class="n">off_atom_index</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">off_atom_coords</span>
                <span class="n">all_zeros</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">all_zeros</span> <span class="ow">and</span> <span class="n">n_atoms</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">molecule</span><span class="o">.</span><span class="n">_add_conformer</span><span class="p">(</span><span class="n">Quantity</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">))</span>

        <span class="c1"># Store charges with implicit units in this scope</span>
        <span class="n">unitless_charges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="n">n_atoms</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>

        <span class="c1"># If all OEAtoms have a partial charge of NaN, then the OFFMol should</span>
        <span class="c1"># have its partial_charges attribute set to None</span>
        <span class="n">any_partial_charge_is_not_nan</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">for</span> <span class="n">oe_atom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="n">oe_idx</span> <span class="o">=</span> <span class="n">oe_atom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>
            <span class="n">off_idx</span> <span class="o">=</span> <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">]</span>
            <span class="n">unitless_charge</span> <span class="o">=</span> <span class="n">oe_atom</span><span class="o">.</span><span class="n">GetPartialCharge</span><span class="p">()</span>
            <span class="c1"># Once this is True, skip the isnancheck</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">any_partial_charge_is_not_nan</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">math</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">unitless_charge</span><span class="p">):</span>
                    <span class="n">any_partial_charge_is_not_nan</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">unitless_charges</span><span class="p">[</span><span class="n">off_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">unitless_charge</span>

        <span class="k">if</span> <span class="n">any_partial_charge_is_not_nan</span><span class="p">:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span> <span class="o">=</span> <span class="n">Quantity</span><span class="p">(</span>
                <span class="n">unitless_charges</span><span class="p">,</span> <span class="n">unit</span><span class="o">.</span><span class="n">elementary_charge</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">return</span> <span class="n">molecule</span>

    <span class="nd">@cached</span><span class="p">(</span><span class="n">LRUCache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="mi">4096</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="n">_mol_to_ctab_and_aro_key</span><span class="p">)</span>
    <span class="k">def</span> <span class="nf">_connection_table_to_openeye</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">,</span> <span class="n">aromaticity_model</span><span class="o">=</span><span class="n">DEFAULT_AROMATICITY_MODEL</span>
    <span class="p">):</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="k">if</span> <span class="n">aromaticity_model</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ALLOWED_AROMATICITY_MODELS</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">InvalidAromaticityModelError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Given aromaticity model </span><span class="si">{</span><span class="n">aromaticity_model</span><span class="si">}</span><span class="s2"> which is not in the set of allowed aromaticity models: &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ALLOWED_AROMATICITY_MODELS</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">()</span>
        <span class="c1"># Add atoms</span>
        <span class="n">off_to_oe_idx</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># {off_idx : oe_idx}</span>
        <span class="n">oemol_atoms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>  <span class="c1"># list of corresponding oemol atoms</span>
        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">molecule</span><span class="o">.</span><span class="n">atoms</span><span class="p">:</span>
            <span class="n">oeatom</span> <span class="o">=</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NewAtom</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">atomic_number</span><span class="p">)</span>
            <span class="n">oeatom</span><span class="o">.</span><span class="n">SetFormalCharge</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">formal_charge</span><span class="o">.</span><span class="n">m_as</span><span class="p">(</span><span class="n">unit</span><span class="o">.</span><span class="n">elementary_charge</span><span class="p">))</span>
            <span class="c1"># TODO: Do we want to provide _any_ pathway for Atom.is_aromatic to influence the OEMol?</span>
            <span class="c1"># oeatom.SetAromatic(atom.is_aromatic)</span>
            <span class="n">oemol_atoms</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oeatom</span><span class="p">)</span>
            <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">atom</span><span class="o">.</span><span class="n">molecule_atom_index</span><span class="p">]</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>

        <span class="c1"># Add bonds</span>
        <span class="n">oemol_bonds</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>  <span class="c1"># list of corresponding oemol bonds</span>
        <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">molecule</span><span class="o">.</span><span class="n">bonds</span><span class="p">:</span>
            <span class="n">atom1_index</span> <span class="o">=</span> <span class="n">bond</span><span class="o">.</span><span class="n">atom1_index</span>
            <span class="n">atom2_index</span> <span class="o">=</span> <span class="n">bond</span><span class="o">.</span><span class="n">atom2_index</span>
            <span class="n">oebond</span> <span class="o">=</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NewBond</span><span class="p">(</span><span class="n">oemol_atoms</span><span class="p">[</span><span class="n">atom1_index</span><span class="p">],</span> <span class="n">oemol_atoms</span><span class="p">[</span><span class="n">atom2_index</span><span class="p">])</span>
            <span class="n">oebond</span><span class="o">.</span><span class="n">SetOrder</span><span class="p">(</span><span class="n">bond</span><span class="o">.</span><span class="n">bond_order</span><span class="p">)</span>
            <span class="c1"># TODO: Do we want to provide _any_ pathway for Bond.is_aromatic to influence the OEMol?</span>
            <span class="c1"># oebond.SetAromatic(bond.is_aromatic)</span>
            <span class="n">oemol_bonds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">oebond</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">aromaticity_model</span> <span class="o">==</span> <span class="s2">&quot;OEAroModel_MDL&quot;</span><span class="p">:</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEAssignAromaticFlags</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAroModelMDL</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">InvalidAromaticityModelError</span><span class="p">(</span>
                <span class="s2">&quot;Aromaticity model </span><span class="si">{aromaticity_model}</span><span class="s2"> is not in the set of allowed aromaticity models:  &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ALLOWED_AROMATICITY_MODELS</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Set atom stereochemistry now that all connectivity is in place</span>
        <span class="k">for</span> <span class="n">atom</span><span class="p">,</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">atoms</span><span class="p">,</span> <span class="n">oemol_atoms</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">atom</span><span class="o">.</span><span class="n">stereochemistry</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="c1"># Set arbitrary initial stereochemistry</span>
            <span class="n">neighs</span> <span class="o">=</span> <span class="p">[</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()]</span>
            <span class="n">oeatom</span><span class="o">.</span><span class="n">SetStereo</span><span class="p">(</span>
                <span class="n">neighs</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomStereo_Tetra</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomStereo_Right</span>
            <span class="p">)</span>

            <span class="c1"># Flip chirality if stereochemistry is incorrect</span>
            <span class="n">oeatom_stereochemistry</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_openeye_cip_atom_stereochemistry</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oeatom</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">oeatom_stereochemistry</span> <span class="o">!=</span> <span class="n">atom</span><span class="o">.</span><span class="n">stereochemistry</span><span class="p">:</span>
                <span class="c1"># Flip the stereochemistry</span>
                <span class="n">oeatom</span><span class="o">.</span><span class="n">SetStereo</span><span class="p">(</span>
                    <span class="n">neighs</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomStereo_Tetra</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomStereo_Left</span>
                <span class="p">)</span>
                <span class="c1"># Verify it matches now as a sanity check</span>
                <span class="n">oeatom_stereochemistry</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_openeye_cip_atom_stereochemistry</span><span class="p">(</span>
                        <span class="n">oemol</span><span class="p">,</span> <span class="n">oeatom</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">oeatom_stereochemistry</span> <span class="o">!=</span> <span class="n">atom</span><span class="o">.</span><span class="n">stereochemistry</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">InconsistentStereochemistryError</span><span class="p">(</span>
                        <span class="s2">&quot;Programming error: OpenEye atom stereochemistry assumptions failed. &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;The atom in the oemol has stereochemistry </span><span class="si">{</span><span class="n">oeatom_stereochemistry</span><span class="si">}</span><span class="s2"> and &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;the atom in the offmol has stereochemistry </span><span class="si">{</span><span class="n">atom</span><span class="o">.</span><span class="n">stereochemistry</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>

        <span class="c1"># Set bond stereochemistry</span>
        <span class="k">for</span> <span class="n">bond</span><span class="p">,</span> <span class="n">oebond</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">bonds</span><span class="p">,</span> <span class="n">oemol_bonds</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">bond</span><span class="o">.</span><span class="n">stereochemistry</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">atom1_index</span> <span class="o">=</span> <span class="n">bond</span><span class="o">.</span><span class="n">molecule</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">bond</span><span class="o">.</span><span class="n">atom1</span><span class="p">)</span>
            <span class="n">atom2_index</span> <span class="o">=</span> <span class="n">bond</span><span class="o">.</span><span class="n">molecule</span><span class="o">.</span><span class="n">atoms</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">bond</span><span class="o">.</span><span class="n">atom2</span><span class="p">)</span>
            <span class="c1"># Set arbitrary initial stereochemistry</span>
            <span class="n">oeatom1</span><span class="p">,</span> <span class="n">oeatom2</span> <span class="o">=</span> <span class="n">oemol_atoms</span><span class="p">[</span><span class="n">atom1_index</span><span class="p">],</span> <span class="n">oemol_atoms</span><span class="p">[</span><span class="n">atom2_index</span><span class="p">]</span>
            <span class="n">oeatom1_neighbor</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">oeatom1</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">n</span> <span class="o">==</span> <span class="n">oeatom2</span><span class="p">)</span>
            <span class="n">oeatom2_neighbor</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">n</span> <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">oeatom2</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">n</span> <span class="o">==</span> <span class="n">oeatom1</span><span class="p">)</span>
            <span class="c1"># oebond.SetStereo([oeatom1, oeatom2], oechem.OEBondStereo_CisTrans, oechem.OEBondStereo_Cis)</span>
            <span class="n">oebond</span><span class="o">.</span><span class="n">SetStereo</span><span class="p">(</span>
                <span class="p">[</span><span class="n">oeatom1_neighbor</span><span class="p">,</span> <span class="n">oeatom2_neighbor</span><span class="p">],</span>
                <span class="n">oechem</span><span class="o">.</span><span class="n">OEBondStereo_CisTrans</span><span class="p">,</span>
                <span class="n">oechem</span><span class="o">.</span><span class="n">OEBondStereo_Cis</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="c1"># Flip stereochemistry if incorrect</span>
            <span class="n">oebond_stereochemistry</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_openeye_cip_bond_stereochemistry</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oebond</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">oebond_stereochemistry</span> <span class="o">!=</span> <span class="n">bond</span><span class="o">.</span><span class="n">stereochemistry</span><span class="p">:</span>
                <span class="c1"># Flip the stereochemistry</span>
                <span class="n">oebond</span><span class="o">.</span><span class="n">SetStereo</span><span class="p">(</span>
                    <span class="p">[</span><span class="n">oeatom1_neighbor</span><span class="p">,</span> <span class="n">oeatom2_neighbor</span><span class="p">],</span>
                    <span class="n">oechem</span><span class="o">.</span><span class="n">OEBondStereo_CisTrans</span><span class="p">,</span>
                    <span class="n">oechem</span><span class="o">.</span><span class="n">OEBondStereo_Trans</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="c1"># Verify it matches now as a sanity check</span>
                <span class="n">oebond_stereochemistry</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_openeye_cip_bond_stereochemistry</span><span class="p">(</span>
                        <span class="n">oemol</span><span class="p">,</span> <span class="n">oebond</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">if</span> <span class="n">oebond_stereochemistry</span> <span class="o">!=</span> <span class="n">bond</span><span class="o">.</span><span class="n">stereochemistry</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="n">InconsistentStereochemistryError</span><span class="p">(</span>
                        <span class="s2">&quot;Programming error: OpenEye bond stereochemistry assumptions failed. &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;The bond in the oemol has stereochemistry </span><span class="si">{</span><span class="n">oebond_stereochemistry</span><span class="si">}</span><span class="s2"> and &quot;</span>
                        <span class="sa">f</span><span class="s2">&quot;the bond in the offmol has stereochemistry </span><span class="si">{</span><span class="n">bond</span><span class="o">.</span><span class="n">stereochemistry</span><span class="si">}</span><span class="s2">.&quot;</span>
                    <span class="p">)</span>

        <span class="c1"># Clean Up phase</span>
        <span class="c1"># The only feature of a molecule that wasn&#39;t perceived above seemed to be ring connectivity, better to run it</span>
        <span class="c1"># here then for someone to inquire about ring sizes and get 0 when it shouldn&#39;t be</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEFindRingAtomsAndBonds</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">oemol</span><span class="p">,</span> <span class="n">off_to_oe_idx</span>

    <span class="k">def</span> <span class="nf">to_openeye</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">,</span>
        <span class="n">aromaticity_model</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">DEFAULT_AROMATICITY_MODEL</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create an OpenEye molecule using the specified aromaticity model</span>

<span class="sd">        ``OEAtom`` s have a different set of allowed value for partial</span>
<span class="sd">        charges than ``openff.toolkit.topology.Molecule``\ s. In the</span>
<span class="sd">        OpenEye toolkits, partial charges are stored on individual</span>
<span class="sd">        ``OEAtom``\ s, and their values are initialized to ``0.0``. In</span>
<span class="sd">        the Open Force Field Toolkit, an``openff.toolkit.topology.Molecule``&#39;s</span>
<span class="sd">        ``partial_charges`` attribute is initialized to ``None`` and can</span>
<span class="sd">        be set to a unit-wrapped numpy array with</span>
<span class="sd">        units of elementary charge. The Open Force Field Toolkit</span>
<span class="sd">        considers an ``OEMol`` where every ``OEAtom`` has a partial</span>
<span class="sd">        charge of ``float(&#39;nan&#39;)`` to be equivalent to an Open Force</span>
<span class="sd">        Field Toolkit ``Molecule``&#39;s ``partial_charges = None``. This</span>
<span class="sd">        assumption is made in both ``to_openeye`` and ``from_openeye``.</span>

<span class="sd">        .. todo ::</span>

<span class="sd">           * Should the aromaticity model be specified in some other way?</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to convert to an OEMol</span>
<span class="sd">        aromaticity_model</span>
<span class="sd">            The aromaticity model to use</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        oemol</span>
<span class="sd">            An OpenEye molecule</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        Create an OpenEye molecule from a Molecule</span>

<span class="sd">        &gt;&gt;&gt; from openff.toolkit import Molecule</span>
<span class="sd">        &gt;&gt;&gt; toolkit_wrapper = OpenEyeToolkitWrapper()</span>
<span class="sd">        &gt;&gt;&gt; molecule = Molecule.from_smiles(&#39;CC&#39;)</span>
<span class="sd">        &gt;&gt;&gt; oemol = toolkit_wrapper.to_openeye(molecule)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span><span class="p">,</span> <span class="n">off_to_oe_idx</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_connection_table_to_openeye</span><span class="p">(</span>
            <span class="n">molecule</span><span class="p">,</span> <span class="n">aromaticity_model</span><span class="o">=</span><span class="n">aromaticity_model</span>
        <span class="p">)</span>
        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="c1"># if not(molecule.name is None):</span>
        <span class="n">oe_to_off_idx</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([(</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">off_to_oe_idx</span><span class="o">.</span><span class="n">items</span><span class="p">()])</span>

        <span class="n">oemol</span><span class="o">.</span><span class="n">SetTitle</span><span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="c1"># Make lists of OE atoms and OE bonds in the same order as the OFF atoms and OFF bonds</span>
        <span class="n">oemol_atoms</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">molecule</span><span class="o">.</span><span class="n">n_atoms</span>  <span class="c1"># list of corresponding oemol atoms</span>
        <span class="k">for</span> <span class="n">oe_atom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="n">oe_idx</span> <span class="o">=</span> <span class="n">oe_atom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>
            <span class="n">oemol_atoms</span><span class="p">[</span><span class="n">oe_to_off_idx</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">]]</span> <span class="o">=</span> <span class="n">oe_atom</span>
            <span class="n">off_atom</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">atoms</span><span class="p">[</span><span class="n">oe_to_off_idx</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">]]</span>
            <span class="n">oe_atom</span><span class="o">.</span><span class="n">SetName</span><span class="p">(</span><span class="n">off_atom</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">off_atom</span><span class="o">.</span><span class="n">partial_charge</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">oe_atom</span><span class="o">.</span><span class="n">SetPartialCharge</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">charge</span> <span class="o">=</span> <span class="n">off_atom</span><span class="o">.</span><span class="n">partial_charge</span><span class="o">.</span><span class="n">m_as</span><span class="p">(</span><span class="n">unit</span><span class="o">.</span><span class="n">elementary_charge</span><span class="p">)</span>
                <span class="n">oe_atom</span><span class="o">.</span><span class="n">SetPartialCharge</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">charge</span><span class="p">))</span>
            <span class="n">res</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomGetResidue</span><span class="p">(</span><span class="n">oe_atom</span><span class="p">)</span>
            <span class="c1"># If we add residue info without updating the serial number, all of the atom</span>
            <span class="c1"># serial numbers in a written PDB will be 0. Note two things:</span>
            <span class="c1"># 1) the &quot;res&quot; object is specific to this atom, so its serial number</span>
            <span class="c1">#    applies only to this atom</span>
            <span class="c1"># 2) we do NOT preserve</span>
            <span class="c1">#    PDB serial numbers in our infrastructure, we merely set these to the</span>
            <span class="c1">#    atom index in the molecule so that OpenEye-written PDBs have</span>
            <span class="c1">#    nonzero atom serial numbers.</span>
            <span class="n">res</span><span class="o">.</span><span class="n">SetSerialNumber</span><span class="p">(</span><span class="n">oe_to_off_idx</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;residue_name&quot;</span> <span class="ow">in</span> <span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetName</span><span class="p">(</span><span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;residue_name&quot;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetName</span><span class="p">(</span><span class="s2">&quot;UNL&quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;residue_number&quot;</span> <span class="ow">in</span> <span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetResidueNumber</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;residue_number&quot;</span><span class="p">]))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetResidueNumber</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;insertion_code&quot;</span> <span class="ow">in</span> <span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetInsertCode</span><span class="p">(</span><span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;insertion_code&quot;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetInsertCode</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;chain_id&quot;</span> <span class="ow">in</span> <span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetChainID</span><span class="p">(</span><span class="n">off_atom</span><span class="o">.</span><span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;chain_id&quot;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">res</span><span class="o">.</span><span class="n">SetChainID</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">)</span>

            <span class="n">oechem</span><span class="o">.</span><span class="n">OEAtomSetResidue</span><span class="p">(</span><span class="n">oe_atom</span><span class="p">,</span> <span class="n">res</span><span class="p">)</span>

        <span class="k">assert</span> <span class="kc">None</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">oemol_atoms</span>

        <span class="n">oemol_bonds</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="n">molecule</span><span class="o">.</span><span class="n">n_bonds</span>  <span class="c1"># list of corresponding oemol bonds</span>
        <span class="k">for</span> <span class="n">oe_bond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
            <span class="n">at1_off_idx</span> <span class="o">=</span> <span class="n">oe_to_off_idx</span><span class="p">[</span><span class="n">oe_bond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">()]</span>
            <span class="n">at2_off_idx</span> <span class="o">=</span> <span class="n">oe_to_off_idx</span><span class="p">[</span><span class="n">oe_bond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">()]</span>
            <span class="n">off_bond</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">get_bond_between</span><span class="p">(</span><span class="n">at1_off_idx</span><span class="p">,</span> <span class="n">at2_off_idx</span><span class="p">)</span>
            <span class="n">off_bond_idx</span> <span class="o">=</span> <span class="n">off_bond</span><span class="o">.</span><span class="n">molecule_bond_index</span>
            <span class="n">oemol_bonds</span><span class="p">[</span><span class="n">off_bond_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">oe_bond</span>
            <span class="k">if</span> <span class="n">off_bond</span><span class="o">.</span><span class="n">fractional_bond_order</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">oe_bond</span><span class="o">.</span><span class="n">SetData</span><span class="p">(</span><span class="s2">&quot;fractional_bond_order&quot;</span><span class="p">,</span> <span class="n">off_bond</span><span class="o">.</span><span class="n">fractional_bond_order</span><span class="p">)</span>

        <span class="k">assert</span> <span class="kc">None</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">oemol_bonds</span>

        <span class="c1"># Retain conformations, if present</span>
        <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">_conformers</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">oemol</span><span class="o">.</span><span class="n">DeleteConfs</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">conf</span> <span class="ow">in</span> <span class="n">molecule</span><span class="o">.</span><span class="n">_conformers</span><span class="p">:</span>
                <span class="c1"># OE needs a 1 x (3*n_Atoms) double array as input</span>
                <span class="n">flat_coords</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="n">oemol</span><span class="o">.</span><span class="n">NumAtoms</span><span class="p">()</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">oe_idx</span> <span class="ow">in</span> <span class="n">off_to_oe_idx</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">z</span><span class="p">)</span> <span class="o">=</span> <span class="n">conf</span><span class="p">[</span><span class="n">index</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">m_as</span><span class="p">(</span><span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">)</span>
                    <span class="n">flat_coords</span><span class="p">[(</span><span class="mi">3</span> <span class="o">*</span> <span class="n">oe_idx</span><span class="p">)]</span> <span class="o">=</span> <span class="n">x</span>
                    <span class="n">flat_coords</span><span class="p">[(</span><span class="mi">3</span> <span class="o">*</span> <span class="n">oe_idx</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">y</span>
                    <span class="n">flat_coords</span><span class="p">[(</span><span class="mi">3</span> <span class="o">*</span> <span class="n">oe_idx</span><span class="p">)</span> <span class="o">+</span> <span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="n">z</span>

                <span class="n">oecoords</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEFloatArray</span><span class="p">(</span><span class="n">flat_coords</span><span class="p">)</span>
                <span class="n">oemol</span><span class="o">.</span><span class="n">NewConf</span><span class="p">(</span><span class="n">oecoords</span><span class="p">)</span>

        <span class="c1"># Retain charges, if present. All atoms are initialized above with a partial charge of NaN.</span>
        <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">oe_indexed_charges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="n">n_atoms</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
            <span class="k">assert</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span><span class="p">,</span> <span class="s2">&quot;units&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">off_idx</span><span class="p">,</span> <span class="n">charge</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span><span class="p">):</span>
                <span class="n">oe_idx</span> <span class="o">=</span> <span class="n">off_to_oe_idx</span><span class="p">[</span><span class="n">off_idx</span><span class="p">]</span>
                <span class="n">charge_unitless</span> <span class="o">=</span> <span class="n">charge</span><span class="o">.</span><span class="n">m_as</span><span class="p">(</span><span class="n">unit</span><span class="o">.</span><span class="n">elementary_charge</span><span class="p">)</span>
                <span class="n">oe_indexed_charges</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">]</span> <span class="o">=</span> <span class="n">charge_unitless</span>
            <span class="c1"># TODO: This loop below fails if we try to use an &quot;enumerate&quot;-style loop.</span>
            <span class="c1">#  It&#39;s worth investigating whether we make this assumption elsewhere in the codebase, since</span>
            <span class="c1">#  the OE docs may indicate that this sort of usage is a very bad thing to do.</span>
            <span class="c1">#  https://docs.eyesopen.com/toolkits/python/oechemtk/atombondindices.html#indices-for-molecule-lookup-considered-harmful</span>
            <span class="c1"># for oe_idx, oe_atom in enumerate(oemol.GetAtoms()):</span>
            <span class="k">for</span> <span class="n">oe_atom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
                <span class="n">oe_idx</span> <span class="o">=</span> <span class="n">oe_atom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>
                <span class="n">oe_atom</span><span class="o">.</span><span class="n">SetPartialCharge</span><span class="p">(</span><span class="n">oe_indexed_charges</span><span class="p">[</span><span class="n">oe_idx</span><span class="p">])</span>

        <span class="c1"># Retain properties, if present</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">molecule</span><span class="o">.</span><span class="n">properties</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OESetSDData</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">key</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">))</span>

        <span class="k">return</span> <span class="n">oemol</span>

    <span class="k">def</span> <span class="nf">atom_is_in_ring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">atom</span><span class="p">:</span> <span class="s2">&quot;Atom&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return whether or not an atom is in a ring.</span>

<span class="sd">        It is assumed that this atom is in molecule.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        atom</span>
<span class="sd">            The molecule containing the atom of interest</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        is_in_ring</span>
<span class="sd">            Whether or not the atom is in a ring.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        NotAttachedToMoleculeError</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">molecule</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NotAttachedToMoleculeError</span><span class="p">(</span>
                <span class="s2">&quot;This Atom does not belong to a Molecule object&quot;</span>
            <span class="p">)</span>

        <span class="n">molecule</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">molecule</span>
        <span class="n">atom_index</span> <span class="o">=</span> <span class="n">atom</span><span class="o">.</span><span class="n">molecule_atom_index</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">()</span>

        <span class="c1"># Molecule.to_openeye() is guaranteed to preserve atom ordering</span>
        <span class="n">is_in_ring</span> <span class="o">=</span> <span class="p">[</span><span class="o">*</span><span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()][</span><span class="n">atom_index</span><span class="p">]</span><span class="o">.</span><span class="n">IsInRing</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">is_in_ring</span>

    <span class="k">def</span> <span class="nf">bond_is_in_ring</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bond</span><span class="p">:</span> <span class="s2">&quot;Bond&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Return whether or not a bond is in a ring.</span>

<span class="sd">        It is assumed that this atom is in molecule.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        bond</span>
<span class="sd">            The molecule containing the atom of interest</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        is_in_ring</span>
<span class="sd">            Whether or not the bond of index `bond_index` is in a ring</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        NotAttachedToMoleculeError</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">bond</span><span class="o">.</span><span class="n">molecule</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">NotAttachedToMoleculeError</span><span class="p">(</span>
                <span class="s2">&quot;This Bond does not belong to a Molecule object&quot;</span>
            <span class="p">)</span>

        <span class="n">molecule</span> <span class="o">=</span> <span class="n">bond</span><span class="o">.</span><span class="n">molecule</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">()</span>

        <span class="c1"># Molecule.to_openeye() is NOT guaranteed to preserve bond ordering,</span>
        <span class="c1"># so we must look up the corresponding `OEBond` via `OEAtom`s</span>
        <span class="n">oebond</span> <span class="o">=</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBond</span><span class="p">(</span>
            <span class="p">[</span><span class="o">*</span><span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()][</span><span class="n">bond</span><span class="o">.</span><span class="n">atom1_index</span><span class="p">],</span>
            <span class="p">[</span><span class="o">*</span><span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">()][</span><span class="n">bond</span><span class="o">.</span><span class="n">atom2_index</span><span class="p">],</span>
        <span class="p">)</span>

        <span class="n">is_in_ring</span> <span class="o">=</span> <span class="n">oebond</span><span class="o">.</span><span class="n">IsInRing</span><span class="p">()</span>

        <span class="k">return</span> <span class="n">is_in_ring</span>

    <span class="k">def</span> <span class="nf">_get_smiles_flavor</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">isomeric</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">explicit_hydrogens</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="c1"># this sets up the default settings following the old DEFAULT flag</span>
        <span class="c1"># more information on flags can be found here</span>
        <span class="c1"># &lt;https://docs.eyesopen.com/toolkits/python/oechemtk/OEChemConstants/OESMILESFlag.html#OEChem::OESMILESFlag&gt;</span>
        <span class="n">smiles_options</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_Canonical</span>
            <span class="o">|</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_Isotopes</span>
            <span class="o">|</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_RGroups</span>
        <span class="p">)</span>

        <span class="c1"># check if we want an isomeric smiles</span>
        <span class="k">if</span> <span class="n">isomeric</span><span class="p">:</span>
            <span class="c1"># add the atom and bond stereo flags</span>
            <span class="n">smiles_options</span> <span class="o">|=</span> <span class="p">(</span>
                <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_AtomStereo</span> <span class="o">|</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_BondStereo</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">explicit_hydrogens</span><span class="p">:</span>
            <span class="c1"># add the hydrogen flag</span>
            <span class="n">smiles_options</span> <span class="o">|=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_Hydrogens</span>

        <span class="k">return</span> <span class="n">smiles_options</span>

    <span class="k">def</span> <span class="nf">to_smiles</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">isomeric</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">explicit_hydrogens</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">mapped</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Uses the OpenEye toolkit to convert a Molecule into a SMILES string.</span>
<span class="sd">        A partially mapped smiles can also be generated for atoms of interest by supplying an `atom_map` to the</span>
<span class="sd">        properties dictionary.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to convert into a SMILES.</span>
<span class="sd">        isomeric</span>
<span class="sd">            return an isomeric smiles</span>
<span class="sd">        explicit_hydrogens</span>
<span class="sd">            return a smiles string containing all hydrogens explicitly</span>
<span class="sd">        mapped</span>
<span class="sd">            return a explicit hydrogen mapped smiles, the atoms to be mapped can be controlled by supplying an</span>
<span class="sd">            atom map into the properties dictionary. If no mapping is passed all atoms will be mapped in order, else</span>
<span class="sd">            an atom map dictionary from the current atom index to the map id should be supplied with no duplicates.</span>
<span class="sd">            The map ids (values) should start from 0 or 1.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        smiles</span>
<span class="sd">            The SMILES of the input molecule.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="n">smiles_options</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_smiles_flavor</span><span class="p">(</span><span class="n">isomeric</span><span class="p">,</span> <span class="n">explicit_hydrogens</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mapped</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">explicit_hydrogens</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">,</span> <span class="p">(</span>
                <span class="s2">&quot;Mapped smiles require all hydrogens and &quot;</span>
                <span class="s2">&quot;stereochemsitry to be defined to retain order&quot;</span>
            <span class="p">)</span>

            <span class="c1"># if we only want to map specific atoms check for an atom map</span>
            <span class="n">atom_map</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">_properties</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;atom_map&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">atom_map</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># make sure there are no repeated indices</span>
                <span class="n">map_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">atom_map</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">map_ids</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">atom_map</span><span class="p">):</span>
                    <span class="n">atom_map</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">elif</span> <span class="mi">0</span> <span class="ow">in</span> <span class="n">atom_map</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                    <span class="c1"># we need to increment the map index</span>
                    <span class="k">for</span> <span class="n">atom</span><span class="p">,</span> <span class="nb">map</span> <span class="ow">in</span> <span class="n">atom_map</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="n">atom_map</span><span class="p">[</span><span class="n">atom</span><span class="p">]</span> <span class="o">=</span> <span class="nb">map</span> <span class="o">+</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">atom_map</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># now we need to add the atom map to the atoms</span>
                <span class="k">for</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
                    <span class="n">oeatom</span><span class="o">.</span><span class="n">SetMapIdx</span><span class="p">(</span><span class="n">oeatom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="c1"># try to set the atom map</span>
                        <span class="n">map_idx</span> <span class="o">=</span> <span class="n">atom_map</span><span class="p">[</span><span class="n">atom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()]</span>
                        <span class="n">atom</span><span class="o">.</span><span class="n">SetMapIdx</span><span class="p">(</span><span class="n">map_idx</span><span class="p">)</span>
                    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                        <span class="k">continue</span>

            <span class="n">smiles_options</span> <span class="o">|=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESMILESFlag_AtomMaps</span>

        <span class="n">smiles</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OECreateSmiString</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">smiles_options</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">smiles</span>

    <span class="k">def</span> <span class="nf">to_inchi</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span> <span class="n">fixed_hydrogens</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create an InChI string for the molecule using the OpenEye OEChem Toolkit.</span>
<span class="sd">        InChI is a standardised representation that does not capture tautomers</span>
<span class="sd">        unless specified using the fixed hydrogen layer.</span>

<span class="sd">        For information on InChi see here https://iupac.org/who-we-are/divisions/division-details/inchi/</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to convert into a SMILES.</span>

<span class="sd">        fixed_hydrogens</span>
<span class="sd">            If a fixed hydrogen layer should be added to the InChI, if `True` this</span>
<span class="sd">            will produce a non standard specific InChI string of the molecule.</span>

<span class="sd">        Returns</span>
<span class="sd">        --------</span>
<span class="sd">        inchi: str</span>
<span class="sd">            The InChI string of the molecule.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        EmptyInChiError</span>
<span class="sd">            If OEChem failed to generate an InChI for the molecule</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fixed_hydrogens</span><span class="p">:</span>
            <span class="n">opts</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEInChIOptions</span><span class="p">()</span>
            <span class="n">opts</span><span class="o">.</span><span class="n">SetFixedHLayer</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">inchi</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMolToInChI</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">inchi</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMolToSTDInChI</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">inchi</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">EmptyInChiError</span><span class="p">(</span>
                <span class="s2">&quot;OEChem failed to generate an InChI for the molecule.&quot;</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">inchi</span>

    <span class="k">def</span> <span class="nf">to_inchikey</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span> <span class="n">fixed_hydrogens</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create an InChIKey for the molecule using the OpenEye OEChem Toolkit.</span>
<span class="sd">        InChIKey is a standardised representation that does not capture tautomers</span>
<span class="sd">        unless specified using the fixed hydrogen layer.</span>

<span class="sd">        For information on InChi see here https://iupac.org/who-we-are/divisions/division-details/inchi/</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to convert into a SMILES.</span>

<span class="sd">        fixed_hydrogens</span>
<span class="sd">            If a fixed hydrogen layer should be added to the InChI, if `True` this will produce a non standard specific</span>
<span class="sd">            InChI string of the molecule.</span>

<span class="sd">        Returns</span>
<span class="sd">        --------</span>
<span class="sd">        inchi_key</span>
<span class="sd">            The InChIKey representation of the molecule.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        EmptyInChiError</span>
<span class="sd">            If OEChem failed to generate an InChI for the molecule</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">fixed_hydrogens</span><span class="p">:</span>
            <span class="n">opts</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEInChIOptions</span><span class="p">()</span>
            <span class="n">opts</span><span class="o">.</span><span class="n">SetFixedHLayer</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">inchi_key</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMolToInChIKey</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">inchi_key</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMolToSTDInChIKey</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">inchi_key</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">EmptyInChiError</span><span class="p">(</span>
                <span class="s2">&quot;OEChem failed to generate an InChI key for the molecule.&quot;</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">inchi_key</span>

    <span class="k">def</span> <span class="nf">to_iupac</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Generate IUPAC name from Molecule</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to convert into a SMILES.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        iupac_name</span>
<span class="sd">            IUPAC name of the molecule</span>

<span class="sd">        Examples</span>
<span class="sd">        --------</span>

<span class="sd">        &gt;&gt;&gt; from openff.toolkit import Molecule</span>
<span class="sd">        &gt;&gt;&gt; from openff.toolkit.utils import get_data_file_path</span>
<span class="sd">        &gt;&gt;&gt; sdf_filepath = get_data_file_path(&#39;molecules/ethanol.sdf&#39;)</span>
<span class="sd">        &gt;&gt;&gt; molecule = Molecule(sdf_filepath)</span>
<span class="sd">        &gt;&gt;&gt; toolkit = OpenEyeToolkitWrapper()</span>
<span class="sd">        &gt;&gt;&gt; iupac_name = toolkit.to_iupac(molecule)</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oeiupac</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">oeiupac</span><span class="o">.</span><span class="n">OECreateIUPACName</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">canonical_order_atoms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Canonical order the atoms in the molecule using the OpenEye toolkit.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The input molecule</span>

<span class="sd">         Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecule</span>
<span class="sd">            The input molecule, with canonically-indexed atoms and bonds.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OECanonicalOrderAtoms</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OECanonicalOrderBonds</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="c1"># reorder the iterator</span>
        <span class="n">vatm</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">atom</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()</span> <span class="o">!=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEElemNo_H</span><span class="p">:</span>
                <span class="n">vatm</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>
        <span class="n">oemol</span><span class="o">.</span><span class="n">OrderAtoms</span><span class="p">(</span><span class="n">vatm</span><span class="p">)</span>

        <span class="n">vbnd</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">bond</span><span class="o">.</span><span class="n">GetBgn</span><span class="p">()</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()</span> <span class="o">!=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEElemNo_H</span>
                <span class="ow">and</span> <span class="n">bond</span><span class="o">.</span><span class="n">GetEnd</span><span class="p">()</span><span class="o">.</span><span class="n">GetAtomicNum</span><span class="p">()</span> <span class="o">!=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEElemNo_H</span>
            <span class="p">):</span>
                <span class="n">vbnd</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bond</span><span class="p">)</span>
        <span class="n">oemol</span><span class="o">.</span><span class="n">OrderBonds</span><span class="p">(</span><span class="n">vbnd</span><span class="p">)</span>

        <span class="n">oemol</span><span class="o">.</span><span class="n">Sweep</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">bond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">()</span> <span class="o">&gt;</span> <span class="n">bond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">():</span>
                <span class="n">bond</span><span class="o">.</span><span class="n">SwapEnds</span><span class="p">()</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">from_smiles</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">smiles</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">hydrogens_are_explicit</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a Molecule from a SMILES string using the OpenEye toolkit.</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        smiles</span>
<span class="sd">            The SMILES string to turn into a molecule</span>
<span class="sd">        hydrogens_are_explicit</span>
<span class="sd">            If False, OE will perform hydrogen addition using OEAddExplicitHydrogens</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            Whether to accept SMILES with undefined stereochemistry. If False,</span>
<span class="sd">            an exception will be raised if a SMILES with undefined stereochemistry</span>
<span class="sd">            is passed into this function.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>
<span class="sd">        name</span>
<span class="sd">            An optional name for the output molecule</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecule</span>
<span class="sd">            An OpenFF style molecule.</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        RadicalsNotSupportedError</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGraphMol</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OESmilesToMol</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">smiles</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">SMILESParseError</span><span class="p">(</span><span class="s2">&quot;Unable to parse the SMILES string&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">hydrogens_are_explicit</span><span class="p">):</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAddExplicitHydrogens</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Addition of explicit hydrogens failed in from_openeye&quot;</span>
                <span class="p">)</span>
        <span class="k">elif</span> <span class="n">hydrogens_are_explicit</span> <span class="ow">and</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEHasImplicitHydrogens</span><span class="p">(</span><span class="n">oemol</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;&#39;hydrogens_are_explicit&#39; was specified as True, but OpenEye Toolkit interpreted &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;SMILES &#39;</span><span class="si">{</span><span class="n">smiles</span><span class="si">}</span><span class="s2">&#39; as having implicit hydrogen. If this SMILES is intended to &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;express all explicit hydrogens in the molecule, then you should construct the &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;desired molecule as an OEMol (where oechem.OEHasImplicitHydrogens(oemol) returns &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;False), and then use Molecule.from_openeye() to create the desired OFFMol.&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Set partial charges to None, since they couldn&#39;t have been stored in a SMILES</span>
        <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="n">atom</span><span class="o">.</span><span class="n">SetPartialCharge</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">))</span>

        <span class="n">molecule</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span>
            <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span><span class="p">,</span>
            <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="n">allow_undefined_stereo</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">molecule</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="k">return</span> <span class="n">molecule</span>

    <span class="k">def</span> <span class="nf">from_inchi</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">inchi</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Construct a Molecule from a InChI representation</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        inchi</span>
<span class="sd">            The InChI representation of the molecule.</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            Whether to accept InChI with undefined stereochemistry. If False,</span>
<span class="sd">            an exception will be raised if a InChI with undefined stereochemistry</span>
<span class="sd">            is passed into this function.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>
<span class="sd">        name</span>
<span class="sd">            An optional name for the output molecule</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecule</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="c1"># This calls the same functions as OESmilesToMol</span>
        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEGraphMol</span><span class="p">()</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEInChIToMol</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">inchi</span><span class="p">)</span>

        <span class="c1"># try and catch InChI parsing fails</span>
        <span class="c1"># if there are no atoms don&#39;t build the molecule</span>
        <span class="k">if</span> <span class="n">oemol</span><span class="o">.</span><span class="n">NumAtoms</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">InChIParseError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;There was an issue parsing the InChI string (</span><span class="si">{</span><span class="n">inchi</span><span class="si">}</span><span class="s2">), please check and try again.&quot;</span>
            <span class="p">)</span>

        <span class="n">molecule</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span>
            <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="n">allow_undefined_stereo</span><span class="p">,</span>
            <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">molecule</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>

        <span class="k">return</span> <span class="n">molecule</span>

    <span class="k">def</span> <span class="nf">from_iupac</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">iupac_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">allow_undefined_stereo</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="s2">&quot;FrozenMolecule&quot;</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Construct a Molecule from an IUPAC name</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        iupac_name</span>
<span class="sd">            The IUPAC or common name of the molecule.</span>
<span class="sd">        allow_undefined_stereo</span>
<span class="sd">            Whether to accept a molecule name with undefined stereochemistry. If False,</span>
<span class="sd">            an exception will be raised if a molecule name with undefined stereochemistry</span>
<span class="sd">            is passed into this function.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        molecule</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span><span class="p">,</span> <span class="n">oeiupac</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">()</span>
        <span class="n">parsing_result</span> <span class="o">=</span> <span class="n">oeiupac</span><span class="o">.</span><span class="n">OEParseIUPACName</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">iupac_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">parsing_result</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">InvalidIUPACNameError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;OpenEye failed to parse </span><span class="si">{</span><span class="n">iupac_name</span><span class="si">}</span><span class="s2"> as a IUPAC name&quot;</span>
            <span class="p">)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OETriposAtomNames</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAddExplicitHydrogens</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">OpenEyeError</span><span class="p">(</span><span class="s2">&quot;Addition of explicit hydrogens failed in from_iupac&quot;</span><span class="p">)</span>

        <span class="n">molecule</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="n">allow_undefined_stereo</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">_cls</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">molecule</span>

    <span class="k">def</span> <span class="nf">generate_conformers</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">n_conformers</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">rms_cutoff</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Quantity</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">clear_existing</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">make_carboxylic_acids_cis</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generate molecule conformers using OpenEye Omega.</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        .. todo ::</span>

<span class="sd">            * which parameters should we expose? (or can we implement a general system with \*\*kwargs?)</span>
<span class="sd">            * will the coordinates be returned in the OpenFF Molecule&#39;s own indexing system? Or is there a chance that</span>
<span class="sd">              they&#39;ll get reindexed when we convert the input into an OEmol?</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to generate conformers for.</span>
<span class="sd">        n_conformers</span>
<span class="sd">            The maximum number of conformers to generate.</span>
<span class="sd">        rms_cutoff</span>
<span class="sd">            The minimum RMS value at which two conformers are considered redundant and one is deleted.</span>
<span class="sd">            If None, the cutoff is set to 1 Angstrom</span>
<span class="sd">        clear_existing</span>
<span class="sd">            Whether to overwrite existing conformers for the molecule</span>
<span class="sd">        make_carboxylic_acids_cis</span>
<span class="sd">            Guarantee all conformers have exclusively cis carboxylic acid groups (COOH)</span>
<span class="sd">            by rotating the proton in any trans carboxylic acids 180 degrees around the C-O bond.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span> <span class="nn">copy</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oeomega</span>

        <span class="c1"># Copy the molecule and scrub the conformers so that omega HAS to read stereo from graph mol</span>
        <span class="c1"># See https://github.com/openforcefield/openff-toolkit/issues/1152</span>
        <span class="n">mol_copy</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
        <span class="n">mol_copy</span><span class="o">.</span><span class="n">_conformers</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">mol_copy</span><span class="p">)</span>
        <span class="n">omega</span> <span class="o">=</span> <span class="n">oeomega</span><span class="o">.</span><span class="n">OEOmega</span><span class="p">()</span>
        <span class="n">omega</span><span class="o">.</span><span class="n">SetMaxConfs</span><span class="p">(</span><span class="n">n_conformers</span><span class="p">)</span>
        <span class="n">omega</span><span class="o">.</span><span class="n">SetCanonOrder</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">omega</span><span class="o">.</span><span class="n">SetSampleHydrogens</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">omega</span><span class="o">.</span><span class="n">SetEnergyWindow</span><span class="p">(</span><span class="mf">15.0</span><span class="p">)</span>  <span class="c1"># unit?</span>
        <span class="k">if</span> <span class="n">rms_cutoff</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">omega</span><span class="o">.</span><span class="n">SetRMSThreshold</span><span class="p">(</span><span class="mf">1.0</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">omega</span><span class="o">.</span><span class="n">SetRMSThreshold</span><span class="p">(</span><span class="n">rms_cutoff</span><span class="o">.</span><span class="n">m_as</span><span class="p">(</span><span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">))</span>
        <span class="c1"># Don&#39;t generate random stereoisomer if not specified</span>
        <span class="n">omega</span><span class="o">.</span><span class="n">SetStrictStereo</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">status</span> <span class="o">=</span> <span class="n">omega</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">status</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="n">omega</span><span class="o">.</span><span class="n">SetStrictStereo</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">new_status</span> <span class="o">=</span> <span class="n">omega</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">new_status</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">ConformerGenerationError</span><span class="p">(</span>
                    <span class="s2">&quot;OpenEye Omega conformer generation failed&quot;</span>
                <span class="p">)</span>

        <span class="n">molecule2</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">from_openeye</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span> <span class="n">allow_undefined_stereo</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">_cls</span><span class="o">=</span><span class="n">molecule</span><span class="o">.</span><span class="vm">__class__</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="n">clear_existing</span><span class="p">:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">_conformers</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">conformer</span> <span class="ow">in</span> <span class="n">molecule2</span><span class="o">.</span><span class="n">_conformers</span><span class="p">:</span>  <span class="c1"># type: ignore[union-attr]</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">_add_conformer</span><span class="p">(</span><span class="n">conformer</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">make_carboxylic_acids_cis</span><span class="p">:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">_make_carboxylic_acids_cis</span><span class="p">(</span><span class="n">toolkit_registry</span><span class="o">=</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">apply_elf_conformer_selection</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">percentage</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">2.0</span><span class="p">,</span>
        <span class="n">limit</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Applies the `ELF method</span>
<span class="sd">        &lt;https://docs.eyesopen.com/toolkits/python/quacpactk/molchargetheory.html#elf-conformer-selection&gt;`_</span>
<span class="sd">        to select a set of diverse</span>
<span class="sd">        conformers which have minimal electrostatically strongly interacting functional</span>
<span class="sd">        groups from a molecules conformers.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        * The input molecule should have a large set of conformers already</span>
<span class="sd">          generated to select the ELF conformers from.</span>
<span class="sd">        * The selected conformers will be retained in the ``molecule.conformers`` list</span>
<span class="sd">          while unselected conformers will be discarded.</span>
<span class="sd">        * Conformers generated with the OpenEye toolkit often include trans</span>
<span class="sd">          carboxylic acids (COOH). These are unphysical and will be rejected by</span>
<span class="sd">          ``apply_elf_conformer_selection``. If no conformers are selected, try</span>
<span class="sd">          re-running ``generate_conformers`` with the ``make_carboxylic_acids_cis``</span>
<span class="sd">          argument set to ``True``</span>

<span class="sd">        See Also</span>
<span class="sd">        --------</span>
<span class="sd">        RDKitToolkitWrapper.apply_elf_conformer_selection</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule which contains the set of conformers to select from.</span>
<span class="sd">        percentage</span>
<span class="sd">            The percentage of conformers with the lowest electrostatic interaction</span>
<span class="sd">            energies to greedily select from.</span>
<span class="sd">        limit</span>
<span class="sd">            The maximum number of conformers to select.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span><span class="p">,</span> <span class="n">oequacpac</span>

        <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">n_conformers</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span>

        <span class="n">oe_molecule</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">()</span>

        <span class="c1"># Select a subset of the OMEGA generated conformers using the ELF10 method.</span>
        <span class="n">oe_elf_options</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEELFOptions</span><span class="p">()</span>
        <span class="n">oe_elf_options</span><span class="o">.</span><span class="n">SetElfLimit</span><span class="p">(</span><span class="n">limit</span><span class="p">)</span>
        <span class="n">oe_elf_options</span><span class="o">.</span><span class="n">SetPercent</span><span class="p">(</span><span class="n">percentage</span><span class="p">)</span>

        <span class="n">oe_elf</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEELF</span><span class="p">(</span><span class="n">oe_elf_options</span><span class="p">)</span>

        <span class="n">output_stream</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">oeosstream</span><span class="p">()</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OEThrow</span><span class="o">.</span><span class="n">SetOutputStream</span><span class="p">(</span><span class="n">output_stream</span><span class="p">)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEThrow</span><span class="o">.</span><span class="n">Clear</span><span class="p">()</span>

        <span class="n">status</span> <span class="o">=</span> <span class="n">oe_elf</span><span class="o">.</span><span class="n">Select</span><span class="p">(</span><span class="n">oe_molecule</span><span class="p">)</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OEThrow</span><span class="o">.</span><span class="n">SetOutputStream</span><span class="p">(</span><span class="n">oechem</span><span class="o">.</span><span class="n">oeerr</span><span class="p">)</span>

        <span class="n">output_string</span> <span class="o">=</span> <span class="n">output_stream</span><span class="o">.</span><span class="n">str</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;UTF-8&quot;</span><span class="p">)</span>
        <span class="n">output_string</span> <span class="o">=</span> <span class="n">output_string</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;Warning: &quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">output_string</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;^: +&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">output_string</span><span class="p">,</span> <span class="n">flags</span><span class="o">=</span><span class="n">re</span><span class="o">.</span><span class="n">MULTILINE</span><span class="p">)</span>
        <span class="n">output_string</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">$&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">output_string</span><span class="p">)</span>

        <span class="c1"># Check to make sure the call to OE was succesful, and re-route any</span>
        <span class="c1"># non-fatal warnings to the correct logger.</span>
        <span class="k">if</span> <span class="n">output_string</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">status</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">output_string</span><span class="p">)</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="n">status</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="s2">&quot;OpenEye failed to select conformers, but did not return any output. &quot;</span>
                <span class="s2">&quot;This most commonly occurs when the Molecule does not have enough conformers to &quot;</span>
                <span class="s2">&quot;select from. Try calling Molecule.apply_elf_conformer_selection() again after &quot;</span>
                <span class="s2">&quot;running Molecule.generate_conformers() with a much larger value of n_conformers &quot;</span>
                <span class="s2">&quot;or with make_carboxylic_acids_cis=True.&quot;</span>
            <span class="p">)</span>
        <span class="k">elif</span> <span class="n">output_string</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="n">output_string</span><span class="p">)</span>

        <span class="c1"># Extract and store the ELF conformers on the input molecule.</span>
        <span class="n">conformers</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">oe_conformer</span> <span class="ow">in</span> <span class="n">oe_molecule</span><span class="o">.</span><span class="n">GetConfs</span><span class="p">():</span>
            <span class="n">conformer</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">oe_molecule</span><span class="o">.</span><span class="n">NumAtoms</span><span class="p">(),</span> <span class="mi">3</span><span class="p">))</span>

            <span class="k">for</span> <span class="n">atom_index</span><span class="p">,</span> <span class="n">coordinates</span> <span class="ow">in</span> <span class="n">oe_conformer</span><span class="o">.</span><span class="n">GetCoords</span><span class="p">()</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">conformer</span><span class="p">[</span><span class="n">atom_index</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">coordinates</span>

            <span class="n">conformers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Quantity</span><span class="p">(</span><span class="n">conformer</span><span class="p">,</span> <span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">))</span>

        <span class="n">molecule</span><span class="o">.</span><span class="n">_conformers</span> <span class="o">=</span> <span class="n">conformers</span>

    <span class="k">def</span> <span class="nf">assign_partial_charges</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">partial_charge_method</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">use_conformers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="n">Quantity</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">strict_n_conformers</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">normalize_partial_charges</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Compute partial charges with OpenEye quacpac, and assign</span>
<span class="sd">        the new values to the partial_charges attribute.</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        .. todo ::</span>

<span class="sd">           * Should the default be ELF?</span>
<span class="sd">           * Can we expose more charge models?</span>


<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            Molecule for which partial charges are to be computed</span>
<span class="sd">        partial_charge_method</span>
<span class="sd">            The charge model to use. One of [&#39;amberff94&#39;, &#39;mmff&#39;, &#39;mmff94&#39;, &#39;am1-mulliken&#39;, &#39;am1bcc&#39;,</span>
<span class="sd">            &#39;am1bccnosymspt&#39;, &#39;am1bccelf10&#39;, &#39;gasteiger&#39;]</span>
<span class="sd">            If None, &#39;am1-mulliken&#39; will be used.</span>
<span class="sd">        use_conformers</span>
<span class="sd">            shape (n_atoms, 3) and dimension of distance. Optional, default = None</span>
<span class="sd">            Coordinates to use for partial charge calculation. If None, an appropriate number</span>
<span class="sd">            of conformers will be generated.</span>
<span class="sd">        strict_n_conformers</span>
<span class="sd">            Whether to raise an exception if an invalid number of conformers is provided for the</span>
<span class="sd">            given charge method.</span>
<span class="sd">            If this is False and an invalid number of conformers is found, a warning will be raised.</span>
<span class="sd">        normalize_partial_charges</span>
<span class="sd">            Whether to offset partial charges so that they sum to the total formal charge of the molecule.</span>
<span class="sd">            This is used to prevent accumulation of rounding errors when the partial charge generation method has</span>
<span class="sd">            low precision.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        ChargeMethodUnavailableError if the requested charge method can not be handled by this toolkit</span>

<span class="sd">        ChargeCalculationError if the charge method is supported by this toolkit, but fails</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span><span class="p">,</span> <span class="n">oequacpac</span>

        <span class="kn">from</span> <span class="nn">openff.toolkit.topology</span> <span class="kn">import</span> <span class="n">Molecule</span>

        <span class="k">if</span> <span class="n">partial_charge_method</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">partial_charge_method</span> <span class="o">=</span> <span class="s2">&quot;am1-mulliken&quot;</span>

        <span class="n">partial_charge_method</span> <span class="o">=</span> <span class="n">partial_charge_method</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">partial_charge_method</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_supported_charge_methods</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ChargeMethodUnavailableError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;partial_charge_method &#39;</span><span class="si">{</span><span class="n">partial_charge_method</span><span class="si">}</span><span class="s2">&#39; is not available from OpenEyeToolkitWrapper. &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;Available charge methods are </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">supported_charge_methods</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">charge_method</span><span class="p">:</span> <span class="n">_ChargeSettings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_supported_charge_methods</span><span class="p">[</span>
            <span class="n">partial_charge_method</span>
        <span class="p">]</span>

        <span class="k">if</span> <span class="n">_cls</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">_cls</span> <span class="o">=</span> <span class="n">Molecule</span>

        <span class="c1"># Make a temporary copy of the molecule, since we&#39;ll be messing with its conformers</span>
        <span class="n">mol_copy</span> <span class="o">=</span> <span class="n">_cls</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">use_conformers</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">charge_method</span><span class="p">[</span><span class="s2">&quot;rec_confs&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">mol_copy</span><span class="o">.</span><span class="n">_conformers</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">generate_conformers</span><span class="p">(</span>
                    <span class="n">mol_copy</span><span class="p">,</span>
                    <span class="n">n_conformers</span><span class="o">=</span><span class="n">charge_method</span><span class="p">[</span><span class="s2">&quot;rec_confs&quot;</span><span class="p">],</span>
                    <span class="n">rms_cutoff</span><span class="o">=</span><span class="mf">0.25</span> <span class="o">*</span> <span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">,</span>
                    <span class="n">make_carboxylic_acids_cis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="c1"># TODO: What&#39;s a &quot;best practice&quot; RMS cutoff to use here?</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mol_copy</span><span class="o">.</span><span class="n">_conformers</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">conformer</span> <span class="ow">in</span> <span class="n">use_conformers</span><span class="p">:</span>
                <span class="n">mol_copy</span><span class="o">.</span><span class="n">_add_conformer</span><span class="p">(</span><span class="n">conformer</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_check_n_conformers</span><span class="p">(</span>
                <span class="n">mol_copy</span><span class="p">,</span>
                <span class="n">partial_charge_method</span><span class="o">=</span><span class="n">partial_charge_method</span><span class="p">,</span>
                <span class="n">min_confs</span><span class="o">=</span><span class="n">charge_method</span><span class="p">[</span><span class="s2">&quot;min_confs&quot;</span><span class="p">],</span>
                <span class="n">max_confs</span><span class="o">=</span><span class="n">charge_method</span><span class="p">[</span><span class="s2">&quot;max_confs&quot;</span><span class="p">],</span>
                <span class="n">strict_n_conformers</span><span class="o">=</span><span class="n">strict_n_conformers</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="n">mol_copy</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">()</span>

        <span class="n">errfs</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">oeosstream</span><span class="p">()</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEThrow</span><span class="o">.</span><span class="n">SetOutputStream</span><span class="p">(</span><span class="n">errfs</span><span class="p">)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEThrow</span><span class="o">.</span><span class="n">Clear</span><span class="p">()</span>

        <span class="n">oe_charge_method</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">oequacpac</span><span class="p">,</span> <span class="n">charge_method</span><span class="p">[</span><span class="s2">&quot;oe_charge_method&quot;</span><span class="p">])</span>

        <span class="c1"># The OpenFF toolkit has always supported a version of AM1BCC with no geometry optimization</span>
        <span class="c1"># or symmetry correction. So we include this keyword to provide a special configuration of quacpac</span>
        <span class="c1"># if requested.</span>
        <span class="k">if</span> <span class="n">partial_charge_method</span> <span class="o">==</span> <span class="s2">&quot;am1bccnosymspt&quot;</span><span class="p">:</span>
            <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;optimize&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
                <span class="s2">&quot;symmetrize&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="p">}</span>
        <span class="k">elif</span> <span class="n">partial_charge_method</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gasteiger&quot;</span><span class="p">,</span> <span class="s2">&quot;mmff94&quot;</span><span class="p">,</span> <span class="s2">&quot;am1bccelf10&quot;</span><span class="p">]:</span>
            <span class="c1"># symmetrize is implicit in gasteiger and mmff and is already set to True in am1bccelf10</span>
            <span class="n">kwargs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;symmetrize&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>

        <span class="k">if</span> <span class="n">partial_charge_method</span> <span class="o">==</span> <span class="s2">&quot;am1elf10&quot;</span><span class="p">:</span>
            <span class="c1"># special case</span>
            <span class="n">charge_settings</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEELFCharges</span><span class="p">(</span>
                <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAM1Charges</span><span class="p">(</span>
                    <span class="n">optimize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">symmetrize</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="p">),</span>
                <span class="mi">10</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="n">charge_settings</span> <span class="o">=</span> <span class="n">oe_charge_method</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="n">quacpac_status</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAssignCharges</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span>
            <span class="n">charge_settings</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OEThrow</span><span class="o">.</span><span class="n">SetOutputStream</span><span class="p">(</span><span class="n">oechem</span><span class="o">.</span><span class="n">oeerr</span><span class="p">)</span>  <span class="c1"># restoring to original state</span>
        <span class="c1"># This logic handles errors encountered in #34, which can occur when using ELF10 conformer selection</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">quacpac_status</span><span class="p">:</span>
            <span class="n">oe_charge_engine</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAM1Charges</span>
                <span class="k">if</span> <span class="n">partial_charge_method</span> <span class="o">==</span> <span class="s2">&quot;am1elf10&quot;</span>
                <span class="k">else</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAM1BCCCharges</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;SelectElfPop: issue with removing trans COOH conformers&quot;</span> <span class="ow">in</span> <span class="p">(</span>
                <span class="n">errfs</span><span class="o">.</span><span class="n">str</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;UTF-8&quot;</span><span class="p">)</span>
            <span class="p">):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Warning: charge assignment involving ELF10 conformer selection failed due to a &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;known bug (toolkit issue #346). Downgrading to </span><span class="si">{</span><span class="n">oe_charge_engine</span><span class="o">.</span><span class="vm">__name__</span><span class="si">}</span><span class="s2"> &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;charge assignment for this molecule. More information &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;is available at https://github.com/openforcefield/openff-toolkit/issues/346&quot;</span>
                <span class="p">)</span>
                <span class="n">quacpac_status</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAssignCharges</span><span class="p">(</span><span class="n">oemol</span><span class="p">,</span> <span class="n">oe_charge_engine</span><span class="p">())</span>

        <span class="k">if</span> <span class="n">quacpac_status</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">ChargeCalculationError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;Unable to assign charges: </span><span class="si">{</span><span class="n">errfs</span><span class="o">.</span><span class="n">str</span><span class="p">()</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;UTF-8&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span>
            <span class="p">)</span>

        <span class="c1"># Extract and return charges</span>
        <span class="c1"># TODO: Make sure atom mapping remains constant</span>
        <span class="c1"># Extract the list of charges, taking into account possible indexing differences</span>
        <span class="n">charges</span> <span class="o">=</span> <span class="n">Quantity</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="n">oemol</span><span class="o">.</span><span class="n">NumAtoms</span><span class="p">(),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">),</span> <span class="n">unit</span><span class="o">.</span><span class="n">elementary_charge</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">oeatom</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="n">index</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>
            <span class="n">charge</span> <span class="o">=</span> <span class="n">oeatom</span><span class="o">.</span><span class="n">GetPartialCharge</span><span class="p">()</span>
            <span class="n">charge</span> <span class="o">=</span> <span class="n">charge</span> <span class="o">*</span> <span class="n">unit</span><span class="o">.</span><span class="n">elementary_charge</span>
            <span class="n">charges</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="o">=</span> <span class="n">charge</span>

        <span class="n">molecule</span><span class="o">.</span><span class="n">partial_charges</span> <span class="o">=</span> <span class="n">charges</span>

        <span class="k">if</span> <span class="n">normalize_partial_charges</span><span class="p">:</span>
            <span class="n">molecule</span><span class="o">.</span><span class="n">_normalize_partial_charges</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">assign_fractional_bond_orders</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">bond_order_model</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">use_conformers</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">list</span><span class="p">[</span><span class="n">Quantity</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">_cls</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Update and store list of bond orders this molecule. Bond orders are stored on each</span>
<span class="sd">        bond, in the `bond.fractional_bond_order` attribute.</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule to assign wiberg bond orders to</span>
<span class="sd">        bond_order_model</span>
<span class="sd">            The charge model to use. One of [&#39;am1-wiberg&#39;, &#39;am1-wiberg-elf10&#39;,</span>
<span class="sd">            &#39;pm3-wiberg&#39;, &#39;pm3-wiberg-elf10&#39;]. If None, &#39;am1-wiberg&#39; will be used.</span>
<span class="sd">        use_conformers</span>
<span class="sd">            The conformers to use for fractional bond order calculation. If None, an</span>
<span class="sd">            appropriate number of conformers will be generated by an available</span>
<span class="sd">            ToolkitWrapper. If the chosen ``bond_order_model`` is an ELF variant, the ELF</span>
<span class="sd">            conformer selection method will be applied to the provided conformers.</span>
<span class="sd">        _cls</span>
<span class="sd">            Molecule constructor</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span><span class="p">,</span> <span class="n">oequacpac</span>

        <span class="k">if</span> <span class="n">_cls</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="kn">from</span> <span class="nn">openff.toolkit.topology.molecule</span> <span class="kn">import</span> <span class="n">Molecule</span>

            <span class="n">_cls</span> <span class="o">=</span> <span class="n">Molecule</span>

        <span class="c1"># Make a copy since we&#39;ll be messing with this molecule&#39;s conformers</span>
        <span class="n">temp_mol</span> <span class="o">=</span> <span class="n">_cls</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">bond_order_model</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">bond_order_model</span> <span class="o">=</span> <span class="s2">&quot;am1-wiberg&quot;</span>

        <span class="n">bond_order_model</span> <span class="o">=</span> <span class="n">bond_order_model</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>

        <span class="n">is_elf_method</span> <span class="o">=</span> <span class="n">bond_order_model</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;am1-wiberg-elf10&quot;</span><span class="p">,</span> <span class="s2">&quot;pm3-wiberg-elf10&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">use_conformers</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">generate_conformers</span><span class="p">(</span>
                <span class="n">molecule</span><span class="o">=</span><span class="n">temp_mol</span><span class="p">,</span>
                <span class="n">n_conformers</span><span class="o">=</span><span class="mi">1</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">is_elf_method</span> <span class="k">else</span> <span class="mi">500</span><span class="p">,</span>
                <span class="c1"># 0.05 is the recommended RMS when generating a &#39;Dense&#39; amount of</span>
                <span class="c1"># conformers using Omega: https://docs.eyesopen.com/toolkits/python/</span>
                <span class="c1"># omegatk/OEConfGenConstants/OEFragBuilderMode.html.</span>
                <span class="n">rms_cutoff</span><span class="o">=</span><span class="kc">None</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">is_elf_method</span> <span class="k">else</span> <span class="mf">0.05</span> <span class="o">*</span> <span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">,</span>
                <span class="n">make_carboxylic_acids_cis</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">temp_mol</span><span class="o">.</span><span class="n">_conformers</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">for</span> <span class="n">conformer</span> <span class="ow">in</span> <span class="n">use_conformers</span><span class="p">:</span>
                <span class="n">temp_mol</span><span class="o">.</span><span class="n">_add_conformer</span><span class="p">(</span><span class="n">conformer</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">temp_mol</span><span class="o">.</span><span class="n">n_conformers</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;No conformers present in molecule submitted for fractional bond order calculation. Consider &quot;</span>
                <span class="s2">&quot;loading the molecule from a file with geometry already present or running &quot;</span>
                <span class="s2">&quot;molecule.generate_conformers() before calling molecule.compute_wiberg_bond_orders()&quot;</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">is_elf_method</span><span class="p">:</span>
            <span class="c1"># Apply the ELF10 conformer selection method.</span>
            <span class="n">temp_mol</span><span class="o">.</span><span class="n">apply_elf_conformer_selection</span><span class="p">()</span>

        <span class="c1"># Set the options to use when computing the WBOs. This is based on example at</span>
        <span class="c1"># https://docs.eyesopen.com/toolkits/python/quacpactk/examples_summary_wibergbondorders.html</span>
        <span class="n">am1</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAM1</span><span class="p">()</span>

        <span class="n">am1results</span> <span class="o">=</span> <span class="n">oequacpac</span><span class="o">.</span><span class="n">OEAM1Results</span><span class="p">()</span>
        <span class="n">am1options</span> <span class="o">=</span> <span class="n">am1</span><span class="o">.</span><span class="n">GetOptions</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">bond_order_model</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;am1-wiberg&quot;</span><span class="p">):</span>
            <span class="n">am1options</span><span class="o">.</span><span class="n">SetSemiMethod</span><span class="p">(</span><span class="n">oequacpac</span><span class="o">.</span><span class="n">OEMethodType_AM1</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">bond_order_model</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;pm3-wiberg&quot;</span><span class="p">):</span>
            <span class="c1"># TODO: Make sure that modifying am1options actually works</span>
            <span class="n">am1options</span><span class="o">.</span><span class="n">SetSemiMethod</span><span class="p">(</span><span class="n">oequacpac</span><span class="o">.</span><span class="n">OEMethodType_PM3</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Bond order model &#39;</span><span class="si">{</span><span class="n">bond_order_model</span><span class="si">}</span><span class="s2">&#39; is not supported by &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;OpenEyeToolkitWrapper. Supported models are [&#39;am1-wiberg&#39;, &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;&#39;am1-wiberg-elf10&#39;, &#39;pm3-wiberg&#39;, &#39;pm3-wiberg-elf10&#39;].&quot;</span>
            <span class="p">)</span>

        <span class="c1"># Convert the conformers into OE friendly objects to make setting them one</span>
        <span class="c1"># at a time easier.</span>
        <span class="n">oe_conformers</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEFloatArray</span><span class="p">(</span><span class="n">conformer</span><span class="o">.</span><span class="n">m_as</span><span class="p">(</span><span class="n">unit</span><span class="o">.</span><span class="n">angstrom</span><span class="p">)</span><span class="o">.</span><span class="n">flatten</span><span class="p">())</span>
            <span class="k">for</span> <span class="n">conformer</span> <span class="ow">in</span> <span class="n">temp_mol</span><span class="o">.</span><span class="n">conformers</span>
        <span class="p">]</span>

        <span class="n">oemol</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">to_openeye</span><span class="p">(</span><span class="n">temp_mol</span><span class="p">)</span>
        <span class="n">bond_orders</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">oe_conformer</span> <span class="ow">in</span> <span class="n">oe_conformers</span><span class="p">:</span>
            <span class="n">oemol</span><span class="o">.</span><span class="n">DeleteConfs</span><span class="p">()</span>
            <span class="n">oemol</span><span class="o">.</span><span class="n">NewConf</span><span class="p">(</span><span class="n">oe_conformer</span><span class="p">)</span>

            <span class="n">status</span> <span class="o">=</span> <span class="n">am1</span><span class="o">.</span><span class="n">CalcAM1</span><span class="p">(</span><span class="n">am1results</span><span class="p">,</span> <span class="n">oemol</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">status</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">OpenEyeError</span><span class="p">(</span>
                    <span class="s2">&quot;Unable to assign charges (in the process of calculating &quot;</span>
                    <span class="s2">&quot;fractional bond orders)&quot;</span>
                <span class="p">)</span>

            <span class="k">for</span> <span class="n">bond</span> <span class="ow">in</span> <span class="n">oemol</span><span class="o">.</span><span class="n">GetBonds</span><span class="p">():</span>
                <span class="n">bond_orders</span><span class="p">[</span><span class="n">bond</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">am1results</span><span class="o">.</span><span class="n">GetBondOrder</span><span class="p">(</span><span class="n">bond</span><span class="o">.</span><span class="n">GetBgnIdx</span><span class="p">(),</span> <span class="n">bond</span><span class="o">.</span><span class="n">GetEndIdx</span><span class="p">())</span>
                <span class="p">)</span>

        <span class="c1"># TODO: Will bonds always map back to the same index? Consider doing a</span>
        <span class="c1">#       topology mapping.</span>
        <span class="k">for</span> <span class="n">bond_idx</span><span class="p">,</span> <span class="n">conformer_bond_orders</span> <span class="ow">in</span> <span class="n">bond_orders</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1"># Get bond order</span>
            <span class="n">order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">conformer_bond_orders</span><span class="p">)</span>

            <span class="n">mol_bond</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">_bonds</span><span class="p">[</span><span class="n">bond_idx</span><span class="p">]</span>
            <span class="n">mol_bond</span><span class="o">.</span><span class="n">fractional_bond_order</span> <span class="o">=</span> <span class="n">order</span>

    <span class="k">def</span> <span class="nf">get_tagged_smarts_connectivity</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">smarts</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">TTA</span><span class="p">,</span> <span class="o">...</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns a tuple of tuples indicating connectivity between tagged atoms in a SMARTS string. Does not</span>
<span class="sd">        return bond order.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        smarts</span>
<span class="sd">            The tagged SMARTS to analyze</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        unique_tags</span>
<span class="sd">            A sorted tuple of all unique tagged atom map indices.</span>
<span class="sd">        tagged_atom_connectivity</span>
<span class="sd">            A tuple of tuples, where each inner tuple is a pair of tagged atoms</span>
<span class="sd">            (tag_idx_1, tag_idx_2) which are bonded. The inner tuples are ordered</span>
<span class="sd">            smallest-to-largest, and the tuple of tuples is ordered lexically. The</span>
<span class="sd">            return value for an improper torsion would be ((1, 2), (2, 3), (2, 4)).</span>

<span class="sd">        Raises</span>
<span class="sd">        ------</span>
<span class="sd">        SMIRKSParsingError</span>
<span class="sd">            If OpenEye toolkit was unable to parse the provided smirks/tagged smarts</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>

        <span class="kn">from</span> <span class="nn">openff.toolkit.utils.exceptions</span> <span class="kn">import</span> <span class="n">SMIRKSParsingError</span>

        <span class="n">qmol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEQMol</span><span class="p">()</span>
        <span class="n">status</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEParseSmarts</span><span class="p">(</span><span class="n">qmol</span><span class="p">,</span> <span class="n">smarts</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">status</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">SMIRKSParsingError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;OpenEye Toolkit was unable to parse SMIRKS </span><span class="si">{</span><span class="n">smarts</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">_unique_tags</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">_connections</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">at1</span> <span class="ow">in</span> <span class="n">qmol</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">at1</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">_unique_tags</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">at1</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">())</span>
            <span class="k">for</span> <span class="n">at2</span> <span class="ow">in</span> <span class="n">at1</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">at2</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">cxn_to_add</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">at1</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">(),</span> <span class="n">at2</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">()])</span>
                <span class="n">_connections</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">cxn_to_add</span><span class="p">))</span>
        <span class="n">connections</span><span class="p">:</span> <span class="n">TTA</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">_connections</span><span class="p">)))</span>
        <span class="n">unique_tags</span><span class="p">:</span> <span class="n">TTA</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">_unique_tags</span><span class="p">)))</span>
        <span class="k">return</span> <span class="n">unique_tags</span><span class="p">,</span> <span class="n">connections</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_find_smarts_matches</span><span class="p">(</span>
        <span class="n">oemol</span><span class="p">,</span>
        <span class="n">smarts</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">aromaticity_model</span><span class="o">=</span><span class="n">DEFAULT_AROMATICITY_MODEL</span><span class="p">,</span>
        <span class="n">unique</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="o">...</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Find all sets of atoms in the provided OpenEye molecule that match the provided SMARTS string.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        oemol</span>
<span class="sd">            oemol to process with the SMIRKS in order to find matches</span>
<span class="sd">        smarts</span>
<span class="sd">            SMARTS string with any number of sequentially tagged atoms.</span>
<span class="sd">            If there are N tagged atoms numbered 1..N, the resulting matches will be N-tuples of</span>
<span class="sd">            atoms that match the corresponding tagged atoms.</span>
<span class="sd">        aromaticity_model</span>
<span class="sd">            The aromaticity model to use. Only OEAroModel_MDL is supported.</span>
<span class="sd">            Molecule is prepared with this aromaticity model prior to querying.</span>
<span class="sd">        unique</span>
<span class="sd">            If True, only return unique matches. If False, return all matches. This is passed to</span>
<span class="sd">            OpenEye&#39;s ``OESubSearch`` as ``uniquematch``.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        matches</span>
<span class="sd">            matches[index] is an N-tuple of atom numbers from the ``oemol``</span>
<span class="sd">            Matches are returned in no guaranteed order.</span>
<span class="sd">            # TODO: What is returned if no matches are found? An empty list, or None?</span>
<span class="sd">            # TODO: Ensure that SMARTS numbers 1, 2, 3... are rendered into order of returned</span>
<span class="sd">            #    matches indexed by 0, 1, 2...</span>

<span class="sd">        .. notes ::</span>

<span class="sd">           * Raises ``LicenseError`` if valid OpenEye tools license is not found, rather than</span>
<span class="sd">               causing program to terminate</span>
<span class="sd">           * Raises ``ChemicalEnvironmentParsingError`` if ``smarts`` query is malformed</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">openeye</span> <span class="kn">import</span> <span class="n">oechem</span>
        <span class="kn">from</span> <span class="nn">openeye.oechem</span> <span class="kn">import</span> <span class="n">OESubSearch</span>

        <span class="kn">from</span> <span class="nn">openff.toolkit.utils.exceptions</span> <span class="kn">import</span> <span class="n">ChemicalEnvironmentParsingError</span>

        <span class="c1"># Make a copy of molecule so we don&#39;t influence original (probably safer than deepcopy per C Bayly)</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEMol</span><span class="p">(</span><span class="n">oemol</span><span class="p">)</span>
        <span class="c1"># Set up query</span>
        <span class="n">qmol</span> <span class="o">=</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEQMol</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEParseSmarts</span><span class="p">(</span><span class="n">qmol</span><span class="p">,</span> <span class="n">smarts</span><span class="p">):</span>
            <span class="k">raise</span> <span class="n">ChemicalEnvironmentParsingError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;OpenEye could not parse the SMARTS/SMIRKS string &quot;</span><span class="si">{</span><span class="n">smarts</span><span class="si">}</span><span class="s1">&quot;&#39;</span>
            <span class="p">)</span>

        <span class="c1"># OEPrepareSearch will clobber our desired aromaticity model if we don&#39;t sync up mol</span>
        <span class="c1"># and qmol ahead of time.</span>

        <span class="n">oechem</span><span class="o">.</span><span class="n">OEClearAromaticFlags</span><span class="p">(</span><span class="n">qmol</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">aromaticity_model</span> <span class="o">==</span> <span class="s2">&quot;OEAroModel_MDL&quot;</span><span class="p">:</span>
            <span class="n">oechem</span><span class="o">.</span><span class="n">OEAssignAromaticFlags</span><span class="p">(</span><span class="n">qmol</span><span class="p">,</span> <span class="n">oechem</span><span class="o">.</span><span class="n">OEAroModel_MDL</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">InvalidAromaticityModelError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Given aromaticity model </span><span class="si">{</span><span class="n">aromaticity_model</span><span class="si">}</span><span class="s2"> which is not in the set of allowed aromaticity models:  &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ALLOWED_AROMATICITY_MODELS</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="c1"># oechem.OEAssignHybridization(mol)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEAssignHybridization</span><span class="p">(</span><span class="n">qmol</span><span class="p">)</span>

        <span class="c1"># Build list of matches</span>
        <span class="c1"># TODO: The MoleculeImage mapping should preserve ordering of template molecule for equivalent atoms</span>
        <span class="c1">#       and speed matching for larger molecules.</span>
        <span class="n">substructure_search</span> <span class="o">=</span> <span class="n">OESubSearch</span><span class="p">(</span><span class="n">qmol</span><span class="p">)</span>
        <span class="c1"># TODO: max_matches = int(max_matches) if max_matches is not None else 0</span>
        <span class="n">max_matches</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">substructure_search</span><span class="o">.</span><span class="n">SetMaxMatches</span><span class="p">(</span><span class="n">max_matches</span><span class="p">)</span>
        <span class="n">oechem</span><span class="o">.</span><span class="n">OEPrepareSearch</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">substructure_search</span><span class="p">)</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">substructure_search</span><span class="o">.</span><span class="n">Match</span><span class="p">(</span><span class="n">mol</span><span class="p">,</span> <span class="n">unique</span><span class="p">):</span>
            <span class="c1"># Compile list of atom indices that match the pattern tags</span>
            <span class="n">atom_indices</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">matched_atom</span> <span class="ow">in</span> <span class="n">match</span><span class="o">.</span><span class="n">GetAtoms</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">matched_atom</span><span class="o">.</span><span class="n">pattern</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">()</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">atom_indices</span><span class="p">[</span><span class="n">matched_atom</span><span class="o">.</span><span class="n">pattern</span><span class="o">.</span><span class="n">GetMapIdx</span><span class="p">()</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                        <span class="n">matched_atom</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">GetIdx</span><span class="p">()</span>
                    <span class="p">)</span>

            <span class="c1"># Compress into tuple</span>
            <span class="n">matches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="nb">tuple</span><span class="p">(</span><span class="n">atom_indices</span><span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">atom_indices</span><span class="p">)))</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">matches</span>

    <span class="k">def</span> <span class="nf">find_smarts_matches</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">molecule</span><span class="p">:</span> <span class="s2">&quot;Molecule&quot;</span><span class="p">,</span>
        <span class="n">smarts</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">aromaticity_model</span><span class="o">=</span><span class="n">DEFAULT_AROMATICITY_MODEL</span><span class="p">,</span>
        <span class="n">unique</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="o">...</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Find all SMARTS matches for the specified molecule, using the specified aromaticity model.</span>

<span class="sd">        .. warning :: This API is experimental and subject to change.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        molecule</span>
<span class="sd">            The molecule for which all specified SMARTS matches are to be located</span>
<span class="sd">        smarts</span>
<span class="sd">            SMARTS string with optional SMIRKS-style atom tagging</span>
<span class="sd">        aromaticity_model</span>
<span class="sd">            The aromaticity model to use. Only OEAroModel_MDL is supported.</span>
<span class="sd">        unique</span>
<span class="sd">            If True, only return unique matches. If False, return all matches.</span>

<span class="sd">        .. note :: Currently, the only supported ``aromaticity_model`` is ``OEAroModel_MDL``</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">oemol</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_connection_table_to_openeye</span><span class="p">(</span><span class="n">molecule</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_find_smarts_matches</span><span class="p">(</span>
            <span class="n">oemol</span><span class="p">,</span>
            <span class="n">smarts</span><span class="p">,</span>
            <span class="n">aromaticity_model</span><span class="o">=</span><span class="n">aromaticity_model</span><span class="p">,</span>
            <span class="n">unique</span><span class="o">=</span><span class="n">unique</span><span class="p">,</span>
        <span class="p">)</span>


<span class="k">def</span> <span class="nf">requires_openeye_module</span><span class="p">(</span><span class="n">module_name</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">inner_decorator</span><span class="p">(</span><span class="n">function</span><span class="p">):</span>
        <span class="nd">@wraps</span><span class="p">(</span><span class="n">function</span><span class="p">)</span>
        <span class="k">def</span> <span class="nf">wrapper</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">module</span> <span class="o">=</span> <span class="n">importlib</span><span class="o">.</span><span class="n">import_module</span><span class="p">(</span><span class="s2">&quot;openeye.&quot;</span> <span class="o">+</span> <span class="n">module_name</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">OpenEyeImportError</span><span class="p">(</span><span class="s2">&quot;openeye.&quot;</span> <span class="o">+</span> <span class="n">module_name</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">license_func</span> <span class="o">=</span> <span class="n">OpenEyeToolkitWrapper</span><span class="o">.</span><span class="n">_license_functions</span><span class="p">[</span><span class="n">module_name</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">OpenEyeImportError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;we do not currently use </span><span class="si">{</span><span class="n">module_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1"># TODO: Custom exception</span>
            <span class="k">assert</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">module</span><span class="p">,</span> <span class="n">license_func</span><span class="p">)()</span>

            <span class="k">return</span> <span class="n">function</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">wrapper</span>

    <span class="k">return</span> <span class="n">inner_decorator</span>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Hannah Turney. Project structure based on the Computational Molecular Science Python Cookiecutter version 1.1.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>